From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <blake.galbreath@gmail.com>
Date: Mon, 10 May 2021 02:08:53 -0500
Subject: [PATCH] Fix SPIGOT-6278


diff --git a/src/main/java/net/minecraft/world/entity/monster/piglin/EntityPiglin.java b/src/main/java/net/minecraft/world/entity/monster/piglin/EntityPiglin.java
index 1f2f836284c52cf3e7ff295a538f88d95e088761..c8bb46f7331b07364ebbf9fbea9a6d2583c3ccb6 100644
--- a/src/main/java/net/minecraft/world/entity/monster/piglin/EntityPiglin.java
+++ b/src/main/java/net/minecraft/world/entity/monster/piglin/EntityPiglin.java
@@ -200,7 +200,7 @@ public class EntityPiglin extends EntityPiglinAbstract implements ICrossbow {
 
     @Override
     public boolean isTypeNotPersistent(double d0) {
-        return !this.isPersistent();
+        return true || !this.isPersistent(); // Purpur - fix SPIGOT-6278
     }
 
     @Override
