From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Phoenix616 <mail@moep.tv>
Date: Thu, 23 Apr 2020 18:03:57 +0100
Subject: [PATCH] Don't wake up entities when damage event is cancelled


diff --git a/src/main/java/net/minecraft/world/entity/EntityLiving.java b/src/main/java/net/minecraft/world/entity/EntityLiving.java
index 515268d0ce80759835669af16f1aafdb721b5b24..13f660bb60b539cba924b75342f9a4189d7b2d63 100644
--- a/src/main/java/net/minecraft/world/entity/EntityLiving.java
+++ b/src/main/java/net/minecraft/world/entity/EntityLiving.java
@@ -1235,9 +1235,12 @@ public abstract class EntityLiving extends Entity {
         } else if (damagesource.isFire() && this.hasEffect(MobEffects.FIRE_RESISTANCE)) {
             return false;
         } else {
-            if (this.isSleeping() && !this.world.isClientSide) {
-                this.entityWakeup();
-            }
+            // Origami start - Don't wake up entities when damage event is cancelled
+            // moved down into damageEntity0
+            //if (this.isSleeping() && !this.world.isClientSide) {
+            //    this.entityWakeup();
+            //}
+            // Origami end
 
             this.ticksFarFromPlayer = 0;
             float f1 = f;
@@ -2006,6 +2009,11 @@ public abstract class EntityLiving extends Entity {
             if (event.isCancelled()) {
                 return false;
             }
+            // Origami start - Don't wake up entities when damage event is cancelled
+            if (this.isSleeping() && !this.world.isClientSide) {
+                this.entityWakeup();
+            }
+            // Origami end
 
             f = (float) event.getFinalDamage();
 
