From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Paul Sauve <paul@technove.co>
Date: Thu, 18 Feb 2021 13:13:27 -0600
Subject: [PATCH] Skip POI finding if stuck in vehicle


diff --git a/src/main/java/net/minecraft/server/BehaviorFindPosition.java b/src/main/java/net/minecraft/server/BehaviorFindPosition.java
index 8d445e9c0875db6cf45e4d8bcfce7cd3d5094d94..bc13f865689058a3c94ee41b8efb4dce5874004b 100644
--- a/src/main/java/net/minecraft/server/BehaviorFindPosition.java
+++ b/src/main/java/net/minecraft/server/BehaviorFindPosition.java
@@ -45,6 +45,12 @@ public class BehaviorFindPosition extends Behavior<EntityCreature> {
     }
 
     protected boolean a(WorldServer worldserver, EntityCreature entitycreature) {
+        // Airplane start - don't calculate POI if they're trapped in a vehicle
+        Entity vehicle = entitycreature.getVehicle();
+        if (vehicle instanceof EntityBoat || vehicle instanceof EntityMinecartRideable) {
+            return false;
+        }
+        // Airplane end
         if (this.d && entitycreature.isBaby()) {
             return false;
         } else if (this.f == 0L) {
