From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Paul Sauve <paul@technove.co>
Date: Mon, 26 Apr 2021 10:52:56 -0500
Subject: [PATCH] Cache climbing check for activation


diff --git a/src/main/java/net/minecraft/world/entity/EntityLiving.java b/src/main/java/net/minecraft/world/entity/EntityLiving.java
index 59b85551533d332880b154039a5cc3683b17a962..515268d0ce80759835669af16f1aafdb721b5b24 100644
--- a/src/main/java/net/minecraft/world/entity/EntityLiving.java
+++ b/src/main/java/net/minecraft/world/entity/EntityLiving.java
@@ -1790,6 +1790,19 @@ public abstract class EntityLiving extends Entity {
         }
     }
 
+    // Airplane start
+    private boolean cachedIsClimbing = false;
+    private BlockPosition lastClimbingPosition = null;
+
+    public boolean isClimbingCached() {
+        if (!this.getChunkCoordinates().equals(this.lastClimbingPosition)) {
+            this.cachedIsClimbing = this.isClimbing();
+            this.lastClimbingPosition = this.getChunkCoordinates();
+        }
+        return this.cachedIsClimbing;
+    }
+    // Airplane end
+
     public IBlockData ds() {
         return this.world.getType(this.getChunkCoordinates());
     }
diff --git a/src/main/java/org/spigotmc/ActivationRange.java b/src/main/java/org/spigotmc/ActivationRange.java
index 3a76d9d90531ed889a0fc92674bd1719aacdb159..2e22c06584c50e32497a2a65b03f5e96063dadc0 100644
--- a/src/main/java/org/spigotmc/ActivationRange.java
+++ b/src/main/java/org/spigotmc/ActivationRange.java
@@ -315,7 +315,7 @@ public class ActivationRange
         if ( entity instanceof EntityLiving )
         {
             EntityLiving living = (EntityLiving) entity;
-            if ( living.isClimbing() || living.jumping || living.hurtTicks > 0 || living.effects.size() > 0 ) // Paper
+            if ( living.isClimbingCached() || living.jumping || living.hurtTicks > 0 || living.effects.size() > 0 ) // Paper // Airplane - use cached climbing
             {
                 return 1; // Paper
             }
