From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Paul Sauve <paul@technove.co>
Date: Mon, 26 Apr 2021 11:34:29 -0500
Subject: [PATCH] Config to disable method profiler


diff --git a/src/main/java/gg/airplane/AirplaneConfig.java b/src/main/java/gg/airplane/AirplaneConfig.java
index c9e00334655ccd273ba108a541ed1cb0633f69b7..65adf3ceda012c8cfdea675c40e2bb27d34ebac7 100644
--- a/src/main/java/gg/airplane/AirplaneConfig.java
+++ b/src/main/java/gg/airplane/AirplaneConfig.java
@@ -111,4 +111,13 @@ public class AirplaneConfig {
     }
 
 
+    public static boolean disableMethodProfiler;
+
+    private static void miscSettings() {
+        config.setComment("misc", "Settings for things that don't belong elsewhere");
+
+        disableMethodProfiler = config.getBoolean("misc.disable-method-profiler", true);
+    }
+
+
 }
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 7b4b9f54510b3a05aad3f7e50e32ee0bf977244a..2bb00a42492b08036e984d3e1d9a564d4b4226c2 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -2196,7 +2196,11 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
         }
     }
 
+    // Airplane start
+    private final GameProfilerFiller disabledProfiler = net.minecraft.util.profiling.GameProfilerDisabled.a;
     public GameProfilerFiller getMethodProfiler() {
+        if (gg.airplane.AirplaneConfig.disableMethodProfiler) return disabledProfiler;
+        // Airplane end
         return this.methodProfiler;
     }
 
