From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Paul Sauve <paul@technove.co>
Date: Sun, 9 May 2021 18:35:05 -0500
Subject: [PATCH] Make VoxelShapeCollisionEntity lazier


diff --git a/src/main/java/net/minecraft/world/phys/shapes/VoxelShapeCollisionEntity.java b/src/main/java/net/minecraft/world/phys/shapes/VoxelShapeCollisionEntity.java
index 45e6943d141082f510833dc92e41a016de766ca5..dbb6a68eea01a3484f2ac862745b57d1d832b80c 100644
--- a/src/main/java/net/minecraft/world/phys/shapes/VoxelShapeCollisionEntity.java
+++ b/src/main/java/net/minecraft/world/phys/shapes/VoxelShapeCollisionEntity.java
@@ -23,14 +23,16 @@ public class VoxelShapeCollisionEntity implements VoxelShapeCollision {
     };
     private final boolean b;
     private final double c;
-    private final Item d;
-    private final Predicate<FluidType> e;
+    private Item d; // Airplane
+    private Entity entity; // Airplane
+    private Predicate<FluidType> e; // Airplane
 
     protected VoxelShapeCollisionEntity(boolean flag, double d0, Item item, Predicate<FluidType> predicate) {
         this.b = flag;
         this.c = d0;
         this.d = item;
         this.e = predicate;
+        this.entity = null; // Airplane
     }
 
     @Deprecated
@@ -57,19 +59,22 @@ public class VoxelShapeCollisionEntity implements VoxelShapeCollision {
         this(
             entity.by(),
             entity.locY(),
-            entity instanceof EntityLiving ? ((EntityLiving) entity).getItemInMainHand().getItem() : Items.AIR,
-            entity instanceof EntityLiving ? ((EntityLiving) entity)::a : (fluidtype) -> false
+            null, //entity instanceof EntityLiving ? ((EntityLiving) entity).getItemInMainHand().getItem() : Items.AIR, // Airplane - lazy
+            null //entity instanceof EntityLiving ? ((EntityLiving) entity)::a : (fluidtype) -> false
         );
+        this.entity = entity;
         // Airplane end
     }
 
     @Override
     public boolean a(Item item) {
+        if (this.d == null) this.d = this.entity instanceof EntityLiving ? ((EntityLiving) this.entity).getItemInMainHand().getItem() : Items.AIR; // Airplane
         return this.d == item;
     }
 
     @Override
     public boolean a(Fluid fluid, FluidTypeFlowing fluidtypeflowing) {
+        if (this.e == null) return this.entity instanceof EntityLiving && ((EntityLiving) this.entity).a(fluidtypeflowing); // Airplane
         return this.e.test(fluidtypeflowing) && !fluid.getType().a((FluidType) fluidtypeflowing);
     }
 
