From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Paul Sauve <paul@technove.co>
Date: Tue, 1 Jun 2021 18:06:29 -0500
Subject: [PATCH] Patch Paper to use fast item merge raytracing


diff --git a/src/main/java/net/minecraft/world/entity/item/EntityItem.java b/src/main/java/net/minecraft/world/entity/item/EntityItem.java
index dd4997e7ffac4773e01add88efec7c0dcd7b4df0..d05d874d54fdf821429814b7b20a3676d426303c 100644
--- a/src/main/java/net/minecraft/world/entity/item/EntityItem.java
+++ b/src/main/java/net/minecraft/world/entity/item/EntityItem.java
@@ -236,10 +236,16 @@ public class EntityItem extends Entity {
                 if (entityitem.z()) {
                     // Paper Start - Fix items merging through walls
                         if (this.world.paperConfig.fixItemsMergingThroughWalls) {
+                            // Airplane start - fast merging!
+                            /*
                             net.minecraft.world.level.RayTrace rayTrace = new net.minecraft.world.level.RayTrace(this.getPositionVector(), entityitem.getPositionVector(),
                                 net.minecraft.world.level.RayTrace.BlockCollisionOption.COLLIDER, net.minecraft.world.level.RayTrace.FluidCollisionOption.NONE, this);
                             net.minecraft.world.phys.MovingObjectPositionBlock rayTraceResult = world.rayTrace(rayTrace);
                             if (rayTraceResult.getType() == net.minecraft.world.phys.MovingObjectPosition.EnumMovingObjectType.BLOCK) continue;
+                             */
+                            if (world.rayTraceDirect(this.getPositionVector(), entityitem.getPositionVector(), net.minecraft.world.phys.shapes.VoxelShapeCollision.a(this)) ==
+                                net.minecraft.world.phys.MovingObjectPosition.EnumMovingObjectType.BLOCK) continue;
+                            // Airplane end
                         }
                     // Paper End
                     this.a(entityitem);
