From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Paul Sauve <paul@technove.co>
Date: Sat, 19 Dec 2020 19:06:37 -0600
Subject: [PATCH] Expose findClass for profiler


diff --git a/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java b/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java
index d2fa792e182736070a119e96b4ae79337de53d2a..4b484d5c2a605188f55ea8cdc0c2839ee6d850ff 100644
--- a/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java
+++ b/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java
@@ -114,6 +114,7 @@ public final class PluginClassLoader extends URLClassLoader { // Spigot
         return loadClass0(name, resolve, true, true);
     }
 
+    public Class<?> _airplane_findClass(@NotNull String name) throws ClassNotFoundException { return this.classes.get(name); } // Airplane
     Class<?> loadClass0(@NotNull String name, boolean resolve, boolean checkGlobal, boolean checkLibraries) throws ClassNotFoundException {
         try {
             return super.loadClass(name, resolve);
