From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bud Gidiere <sgidiere@gmail.com>
Date: Sat, 1 Aug 2020 15:51:06 -0500
Subject: [PATCH] Yatopia API Bundle

Lagging threshold + Purpur & Rainforest & Origami config

diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 62cc1c74c11f56dcbd1e24e9c5478497742e6351..59787c474f1a17dff156e74107f880ac083a7cb3 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -1776,4 +1776,15 @@ public final class Bukkit {
     public static Server.Spigot spigot() {
         return server.spigot();
     }
+
+    // Purpur start
+    /**
+     * Check if the server is lagging, according to the laggy threshold setting.
+     *
+     * @return true if lagging
+     */
+    public static boolean isLagging() {
+        return server.isLagging();
+    }
+    // Purpur end
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 46c4d98ce32a0b1d8cfd391f2c7a9566a045cb55..3d06d68f41328e18994dc20dc407f60f6b08ada8 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1455,6 +1455,47 @@ public interface Server extends PluginMessageRecipient {
         }
         // Tuinity end - add config to timings report
 
+        // Akarin start
+        /**
+         * @deprecated yatopia does not import akarin config anymore
+         */
+        @Deprecated
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getAkarinConfig()
+        {
+            return new org.bukkit.configuration.file.YamlConfiguration();
+        }
+        // Akarin end
+
+        // Rainforest start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getRainforestConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Rainforest end
+
+        // Purpur start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getPurpurConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+
+        @NotNull
+        public java.util.Properties getServerProperties()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Purpur end
+        // Origami start - add config to timings report
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getOrigamiConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Origami end
+
         /**
          * Sends the component to the player
          *
@@ -1555,4 +1596,13 @@ public interface Server extends PluginMessageRecipient {
     @NotNull
     com.destroystokyo.paper.entity.ai.MobGoals getMobGoals();
     // Paper end
+
+    // Purpur start
+    /**
+     * Check if the server is lagging, according to the laggy threshold setting.
+     *
+     * @return true if lagging
+     */
+    boolean isLagging();
+    // Purpur end
 }
