From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bud Gidiere <sgidiere@gmail.com>
Date: Sat, 1 Aug 2020 15:29:16 -0500
Subject: [PATCH] Yatopia API Fixes


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 62cc1c74c11f56dcbd1e24e9c5478497742e6351..e40542f8e72d9ee8137c07afc01e5534fd91bd36 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -1776,4 +1776,15 @@ public final class Bukkit {
     public static Server.Spigot spigot() {
         return server.spigot();
     }
+
+    // Purpur Start
+    /**
+     * Check if server is lagging according to laggy threshold setting
+     *
+     * @return True if lagging
+     */
+     public static boolean isLagging() {
+     return server.isLagging();
+     }
+    // Purpur end
 }
diff --git a/src/main/java/org/bukkit/Bukkit.java.rej b/src/main/java/org/bukkit/Bukkit.java.rej
deleted file mode 100644
index ff7a0fef308dc3cd14bf70aa1b93f2bedbc187c6..0000000000000000000000000000000000000000
--- a/src/main/java/org/bukkit/Bukkit.java.rej
+++ /dev/null
@@ -1,16 +0,0 @@
-diff a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java	(rejected hunks)
-@@ -1786,5 +1786,14 @@ public final class Bukkit {
-     public static String getServerName() {
-         return server.getServerName();
-     }
-+
-+    /**
-+     * Check if server is lagging according to laggy threshold setting
-+     *
-+     * @return True if lagging
-+     */
-+    public static boolean isLagging() {
-+        return server.isLagging();
-+    }
-     // Purpur end
- }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 1f916988bd54e8386011b0d7e82fc9219939a881..ebc235e3973ca0b9de5604409e486d9f5f38bbdf 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1463,6 +1463,28 @@ public interface Server extends PluginMessageRecipient {
         }
         // Akarin end - Server config
 
+        // Papercut start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getPapercutConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Papercut end
+
+        // Purpur start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getPurpurConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+
+        @NotNull
+        public java.util.Properties getServerProperties()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Purpur end
+
         /**
          * Sends the component to the player
          *
@@ -1563,4 +1585,11 @@ public interface Server extends PluginMessageRecipient {
     @NotNull
     com.destroystokyo.paper.entity.ai.MobGoals getMobGoals();
     // Paper end
+
+    /**
+     * Check if server is lagging according to laggy threshold setting
+     *
+     * @return True if lagging
+     */
+     boolean isLagging();
 }
diff --git a/src/main/java/org/bukkit/Server.java.rej b/src/main/java/org/bukkit/Server.java.rej
deleted file mode 100644
index 27ac6085245ef4270b20b213f9520b87a6c7b323..0000000000000000000000000000000000000000
--- a/src/main/java/org/bukkit/Server.java.rej
+++ /dev/null
@@ -1,14 +0,0 @@
-diff a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java	(rejected hunks)
-@@ -1575,5 +1575,12 @@ public interface Server extends PluginMessageRecipient {
-      */
-     @NotNull
-     String getServerName();
-+
-+    /**
-+     * Check if server is lagging according to laggy threshold setting
-+     *
-+     * @return True if lagging
-+     */
-+    boolean isLagging();
-     // Purpur end
- }
