From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: YatopiaMC <yatopiamc@gmail.com>
Date: Sat, 1 Aug 2020 15:51:06 -0500
Subject: [PATCH] Yatopia API Bundle

Lagging threshold + Purpur & Rainforest & Origami config

diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 464b2161f42a9a3969336820bc793274f4d6e942..a1c7fdcf34519c4781fcd8190d7c2ee787448694 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -1808,4 +1808,15 @@ public final class Bukkit {
     public static Server.Spigot spigot() {
         return server.spigot();
     }
+
+    // Purpur start
+    /**
+     * Check if the server is lagging, according to the laggy threshold setting.
+     *
+     * @return true if lagging
+     */
+    public static boolean isLagging() {
+        return server.isLagging();
+    }
+    // Purpur end
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 7a51edbde474bf610cb5928de4b1bbe2edf34638..720b39459382511cf23ee8af431f0fc4a691358c 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1483,6 +1483,58 @@ public interface Server extends PluginMessageRecipient {
         }
         // Tuinity end - add config to timings report
 
+        // Akarin start
+        /**
+         * @deprecated yatopia does not import akarin config anymore
+         */
+        @Deprecated
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getAkarinConfig()
+        {
+            return new org.bukkit.configuration.file.YamlConfiguration();
+        }
+        // Akarin end
+
+        // Rainforest start
+        /**
+         * @deprecated yatopia does not import akarin config anymore
+         */
+        @Deprecated
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getRainforestConfig()
+        {
+            return new org.bukkit.configuration.file.YamlConfiguration();
+        }
+        // Rainforest end
+
+        // Purpur start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getPurpurConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+
+        @NotNull
+        public java.util.Properties getServerProperties()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Purpur end
+        // Origami start - add config to timings report
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getOrigamiConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Origami end
+        // Yatopia start - add config to timings report
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getYatopiaConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Yatopia end
+
         /**
          * Sends the component to the player
          *
@@ -1583,4 +1635,13 @@ public interface Server extends PluginMessageRecipient {
     @NotNull
     com.destroystokyo.paper.entity.ai.MobGoals getMobGoals();
     // Paper end
+
+    // Purpur start
+    /**
+     * Check if the server is lagging, according to the laggy threshold setting.
+     *
+     * @return true if lagging
+     */
+    boolean isLagging();
+    // Purpur end
 }
