From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: YatopiaMC <yatopiamc@gmail.com>
Date: Sun, 17 Jan 2021 15:37:52 -0600
Subject: [PATCH] Yatopia Config & Redirect Config


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index ea140aaf20f5d4212c109a2bfc0aab1e0851c665..8e7527a848a8434fac5a1b0567ce20ac4c251449 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1607,6 +1607,22 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Purpur end
 
+        // Origami start - add config to timings report
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getOrigamiConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Origami end
+
+        // Yatopia start - add config to timings report
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getYatopiaConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Yatopia end
+
         /**
          * Sends the component to the player
          *
