From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Toffikk <slodkitofik@gmail.com>
Date: Thu, 17 Jun 2021 14:44:11 +0200
Subject: [PATCH] Fix Javadoc generation


diff --git a/build.gradle.kts b/build.gradle.kts
index 61686eaa716cc07fdfad6903a543fdf3b635b3f6..e90ecf61f68a864f5542a418bf8798539c7631e4 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -69,6 +69,8 @@ tasks.jar {
 }
 
 tasks.withType<Javadoc>().configureEach {
+    (options as StandardJavadocDocletOptions).encoding = "UTF-8"
+    (options as StandardJavadocDocletOptions).addBooleanOption("html5", true)
     (options as StandardJavadocDocletOptions).links(
         "https://guava.dev/releases/21.0/api/docs/",
         "https://javadoc.io/doc/org.yaml/snakeyaml/1.27/",
