From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Toffikk <slodkitofik@gmail.com>
Date: Thu, 17 Jun 2021 14:44:11 +0200
Subject: [PATCH] Fix Javadoc generation


diff --git a/build.gradle.kts b/build.gradle.kts
index 6d04816e22f44a33c001d2b7e080402fba6af86c..24820ac11fd70d4f0db3fd7d51a4e48fbd6ee184 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -68,6 +68,8 @@ tasks.jar {
 }
 
 tasks.withType<Javadoc>().configureEach {
+    (options as StandardJavadocDocletOptions).encoding = "UTF-8"
+    (options as StandardJavadocDocletOptions).addBooleanOption("html5", true)
     (options as StandardJavadocDocletOptions).links(
         "https://guava.dev/releases/21.0/api/docs/",
         "https://javadoc.io/doc/org.yaml/snakeyaml/1.27/",
