From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: tr7zw <tr7zw@live.de>
Date: Thu, 5 Mar 2020 21:13:14 +0100
Subject: [PATCH] Add getLastTickMs() api


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index af770b69d2096c8f693bb0e4988b3d84b66ce38e..eaadd170b22f686c8e87e58fe45c146310403676 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -1772,6 +1772,14 @@ public final class Bukkit {
     }
     // Paper end
 
+    // YAPFA start
+
+    public static long getLastTickMs() {
+	return server.getLastTickMs();
+    }
+
+    // YAPFA end
+
     @NotNull
     public static Server.Spigot spigot() {
         return server.spigot();
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 859efc5c5f50d48ca1557056da659ed182d5c181..dd82b79465af847813a4b73fa812337c375827c4 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1589,4 +1589,15 @@ public interface Server extends PluginMessageRecipient {
     @NotNull
     com.destroystokyo.paper.entity.ai.MobGoals getMobGoals();
     // Paper end
+
+    // YAPFA start
+
+    /**
+     * Get the ms the last servertick took
+     *
+     * @return Last ticks ms
+     */
+    long getLastTickMs();
+
+    // YAPFA end
 }
