From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: tr7zw <tr7zw@live.de>
Date: Sat, 1 Aug 2020 15:54:43 -0500
Subject: [PATCH] Add getLastTickMs api


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 59787c474f1a17dff156e74107f880ac083a7cb3..74bb5aa790d072c130a05d0e59afdb1cbf359440 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -1787,4 +1787,10 @@ public final class Bukkit {
         return server.isLagging();
     }
     // Purpur end
+
+    // Yatopia start
+    public static long getLastTickMs() {
+        return server.getLastTickMs();
+    }
+    // Yatopia end
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 3d06d68f41328e18994dc20dc407f60f6b08ada8..87174c63178ac2acd6dfd23d499e4c7efccaa981 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1605,4 +1605,13 @@ public interface Server extends PluginMessageRecipient {
      */
     boolean isLagging();
     // Purpur end
+
+    // Yatopia start
+     /**
+     * Get the ms the last servertick took
+     *
+     * @return Last ticks ms
+     */
+     long getLastTickMs();
+    // Yatopia end
 }
