From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bud Gidiere <sgidiere@gmail.com>
Date: Sat, 1 Aug 2020 15:29:58 -0500
Subject: [PATCH] Add getLastTickMs() api


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index e40542f8e72d9ee8137c07afc01e5534fd91bd36..79276867d2c7e785a2330f91d4d50d27783329f7 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -1772,6 +1772,14 @@ public final class Bukkit {
     }
     // Paper end
 
+    // YAPFA start
+
+    public static long getLastTickMs() {
+	return server.getLastTickMs();
+    }
+
+    // YAPFA end
+
     @NotNull
     public static Server.Spigot spigot() {
         return server.spigot();
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index ebc235e3973ca0b9de5604409e486d9f5f38bbdf..7a5f7a1c24e54b6863033fad3e7ae0163f4e1ff6 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1592,4 +1592,15 @@ public interface Server extends PluginMessageRecipient {
      * @return True if lagging
      */
      boolean isLagging();
+
+    // YAPFA start
+
+    /**
+     * Get the ms the last servertick took
+     *
+     * @return Last ticks ms
+     */
+    long getLastTickMs();
+
+    // YAPFA end
 }
