From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bud Gidiere <sgidiere@gmail.com>
Date: Sat, 1 Aug 2020 15:54:43 -0500
Subject: [PATCH] Add getLastTickMs api


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 0770736b4b52f325027671b43f74ad038c97ce7f..d33e3053e2a7b79f0e5dd699d729797720251a69 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -1787,4 +1787,10 @@ public final class Bukkit {
         return server.isLagging();
     }
     // Purpur end
+
+    // YAPFA start
+    public static long getLastTickMs() {
+        return server.getLastTickMs();
+    }
+    // YAPFA end
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 344db2c4603667edf804bedf2ecae0860f6a672d..3199d6524b79162961badeee30f397bc158f25dd 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1589,4 +1589,13 @@ public interface Server extends PluginMessageRecipient {
     */
     boolean isLagging();
     // Purpur end
+
+    // YAPFA start
+     /**
+     * Get the ms the last servertick took
+     *
+     * @return Last ticks ms
+     */
+     long getLastTickMs();
+    // YAPFA end
 }
