From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: fluse1367 <18399629+fluse1367@users.noreply.github.com>
Date: Sun, 17 Jan 2021 13:58:07 +0100
Subject: [PATCH] Entity Tracking


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index d6ba0895f259c15de9fb974a5c74da709158fc28..8ddb5aedf57a9d64e946d8b9700d8406bca79d79 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1490,6 +1490,14 @@ public interface Server extends PluginMessageRecipient {
         }
         // Tuinity end - add config to timings report
 
+        // Entity Tracking start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getEntityTrackingConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Entity Tracking end
+
         // Akarin start
         /**
          * @deprecated yatopia does not import akarin config anymore
