From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: tr7zw <tr7zw@live.de>
Date: Thu, 2 Apr 2020 18:49:38 +0200
Subject: [PATCH] Add NBT API as a first-class lib


diff --git a/build.gradle.kts b/build.gradle.kts
index 73d15170e361e9f7de3cfc66e4c193cdcc596b89..086426fa989f4dd11d3297b06730f51d8cdbed63 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -103,6 +103,9 @@ tasks.shadowJar {
     relocate("org.bukkit.craftbukkit", "org.bukkit.craftbukkit.v$packageVersion") {
         exclude("org.bukkit.craftbukkit.Main*")
     }
+
+    relocate("de.tr7zw.changeme.nbtapi", "de.tr7zw.nbtapi")
+
     listOf(
         "jline", "org.apache.commons.codec", "org.apache.commons.io", // Paper - don't relocate fastutil
         "org.apache.commons.lang3", "org.apache.http", "org.apache.maven", "org.codehaus.plexus",
