From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: tr7zw <tr7zw@live.de>
Date: Thu, 2 Apr 2020 18:49:38 +0200
Subject: [PATCH] Add NBT API as a first-class lib


diff --git a/build.gradle.kts b/build.gradle.kts
index 80a2ba2fe63ad6fd6e83f46cd29b4cabe64023b0..b68e010579968ac11a336d1c18704eaf18316056 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -109,6 +109,7 @@ tasks.shadowJar {
     relocate("org.bukkit.craftbukkit", "org.bukkit.craftbukkit.v$packageVersion") {
         exclude("org.bukkit.craftbukkit.Main*")
     }
+    relocate("de.tr7zw.changeme.nbtapi", "de.tr7zw.nbtapi")
     mergeServiceFiles()
     transform(ModifiedLog4j2PluginsCacheFileTransformer::class.java)
 }
