From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: tr7zw <tr7zw@live.de>
Date: Fri, 31 Jul 2020 21:28:00 -0500
Subject: [PATCH] Allow to change the piston push limit


diff --git a/src/main/java/de/tr7zw/yatopia/YapfaConfig.java b/src/main/java/de/tr7zw/yatopia/YapfaConfig.java
index ee1cea164ad5c1653502a08b31b1df482a598e14..47434bd82bd5cb8cbe4eac18423c436b9762be14 100644
--- a/src/main/java/de/tr7zw/yatopia/YapfaConfig.java
+++ b/src/main/java/de/tr7zw/yatopia/YapfaConfig.java
@@ -209,5 +209,10 @@ public class YapfaConfig {
     public static boolean onlyPlayerCollisions = false;
     private static void onlyPlayerCollisions() {
         onlyPlayerCollisions = getBoolean("settings.onlyPlayerCollisions", false);
-        }
+    }
+
+    public static int pistonPushLimit = 12;
+    private static void pistonPushLimit() {
+        pistonPushLimit = getInt("settings.pistonPushLimit", 12);
+    }
 }
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/server/PistonExtendsChecker.java b/src/main/java/net/minecraft/server/PistonExtendsChecker.java
index 95aeaaf0bb07c5ecf7117dae3382b0f0aac6426f..b23e96ef11819f2892d8b50937fb68eaf0bc2a63 100644
--- a/src/main/java/net/minecraft/server/PistonExtendsChecker.java
+++ b/src/main/java/net/minecraft/server/PistonExtendsChecker.java
@@ -79,7 +79,7 @@ public class PistonExtendsChecker {
         } else {
             int i = 1;
 
-            if (i + this.f.size() > 12) {
+            if (i + this.f.size() > dev.tr7zw.yatopia.YapfaConfig.pistonPushLimit) { // Yatopia
                 return false;
             } else {
                 while (a(block)) {
@@ -93,7 +93,7 @@ public class PistonExtendsChecker {
                     }
 
                     ++i;
-                    if (i + this.f.size() > 12) {
+                    if (i + this.f.size() > dev.tr7zw.yatopia.YapfaConfig.pistonPushLimit) { // Yatopia
                         return false;
                     }
                 }
@@ -141,7 +141,7 @@ public class PistonExtendsChecker {
                         return true;
                     }
 
-                    if (this.f.size() >= 12) {
+                    if (this.f.size() >= dev.tr7zw.yatopia.YapfaConfig.pistonPushLimit) { // Yatopia
                         return false;
                     }
 
