From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ishland <ishlandmc@yeah.net>
Date: Wed, 27 Jan 2021 20:16:47 +0800
Subject: [PATCH] Preload ProtocolLib EnumWrappers


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 1113777e6e212cb0245141826d11bd3a571216cf..685220edb6690d2d560a7153d5eb9877238b57c0 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1135,6 +1135,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
                 LOGGER.info("Done ({})! For help, type \"help\"", doneTime);
                 // Paper end
 
+                if (org.yatopiamc.yatopia.server.YatopiaConfig.fixProtocolLib) org.yatopiamc.yatopia.server.util.YatopiaPreloadProtocolLib.run(); // Yatopia - Preload ProtocolLib's Enum Wrappers
                 org.spigotmc.WatchdogThread.tick(); // Paper
                 org.spigotmc.WatchdogThread.hasStarted = true; // Paper
                 Arrays.fill( recentTps, 20 );
diff --git a/src/main/java/org/yatopiamc/yatopia/server/YatopiaConfig.java b/src/main/java/org/yatopiamc/yatopia/server/YatopiaConfig.java
index eb6786fa040ede035b649d3f9c2c117afe390406..f3d0daca99a9a80fec99c44c32318e98d162979f 100644
--- a/src/main/java/org/yatopiamc/yatopia/server/YatopiaConfig.java
+++ b/src/main/java/org/yatopiamc/yatopia/server/YatopiaConfig.java
@@ -108,18 +108,11 @@ public class YatopiaConfig {
 			num = 0D;
 		}
 		switch (unit) {
-			case 'd':
-				num *= (double) 60 * 60 * 24;
-				break;
-				case 'h':
-				num *= (double) 60 * 60;
-				break;
-				case 'm':
-				num *= 60;
-				break;
-				default:
-				case 's':
-					break;
+			case 'd':  num *= (double) 60 * 60 * 24;
+			case 'h':  num *= (double) 60 * 60;
+			case 'm':  num *= 60;
+			case 's':  break;
+			default: break;
 		}
 		return (int) num;
 }
@@ -177,5 +171,10 @@ public class YatopiaConfig {
 		return config.getString(path, config.getString(path));
 }
 
+	public static boolean fixProtocolLib = true;
+	private static void protocolLib() {
+		fixProtocolLib = getBoolean("settings.fix-protocollib", fixProtocolLib);
+	}
+
 	public static boolean disableEntityStuckChecks = false;
 private static void disableEntityStuckChecks() {
 	disableEntityStuckChecks = getBoolean("settings.disableEntityStuckChecks", false);
diff --git a/src/main/java/org/yatopiamc/yatopia/server/util/YatopiaPreloadProtocolLib.java b/src/main/java/org/yatopiamc/yatopia/server/util/YatopiaPreloadProtocolLib.java
new file mode 100644
index 0000000000000000000000000000000000000000..85906aa00163a4626b16190e2e48385bc5eba801
--- /dev/null
+++ b/src/main/java/org/yatopiamc/yatopia/server/util/YatopiaPreloadProtocolLib.java
@@ -0,0 +1,28 @@
+package org.yatopiamc.yatopia.server.util;
+
+import net.minecraft.server.MinecraftServer;
+import org.bukkit.Bukkit;
+import org.bukkit.plugin.Plugin;
+import org.bukkit.plugin.SimplePluginManager;
+
+import java.lang.reflect.Method;
+
+public class YatopiaPreloadProtocolLib {
+
+    public synchronized static void run() {
+        try {
+            final SimplePluginManager pluginManager = (SimplePluginManager) Bukkit.getPluginManager();
+            final Plugin protocolLib = pluginManager.getPlugin("ProtocolLib");
+            if(protocolLib != null && protocolLib.isEnabled()) {
+                MinecraftServer.LOGGER.info("Yatopia: Attempting to fix ProtocolLib");
+                final Method initialize = Class.forName("com.comphenix.protocol.wrappers.EnumWrappers", true, protocolLib.getClass().getClassLoader()).getDeclaredMethod("initialize");
+                initialize.setAccessible(true);
+                initialize.invoke(null);
+                synchronized (YatopiaPreloadProtocolLib.class) {
+                }
+            }
+        } catch (Throwable t) {
+            MinecraftServer.LOGGER.warn("Unable to fix ProtocolLib", t);
+        }
+    }
+}
