From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: tr7zw <tr7zw@live.de>
Date: Fri, 31 Jul 2020 22:39:45 -0500
Subject: [PATCH] Player-saving-async-FileIO


diff --git a/src/main/java/net/minecraft/server/AdvancementDataPlayer.java b/src/main/java/net/minecraft/server/AdvancementDataPlayer.java
index cf539c98073b475eb5b769c8cc11d48a7e6d58f1..93b14f01fd7094175968918f0dcd7c3c921f156a 100644
--- a/src/main/java/net/minecraft/server/AdvancementDataPlayer.java
+++ b/src/main/java/net/minecraft/server/AdvancementDataPlayer.java
@@ -50,6 +50,7 @@ public class AdvancementDataPlayer {
     @Nullable
     private Advancement l;
     private boolean m = true;
+    public static java.util.concurrent.ExecutorService saveThread = java.util.concurrent.Executors.newSingleThreadExecutor(); // Yatopia
 
     // Paper start - fix advancement data player leakage
     final Map<CriterionTriggerAbstract, Set<CriterionTrigger.a>> criterionData = Maps.newIdentityHashMap();
@@ -228,6 +229,15 @@ public class AdvancementDataPlayer {
 
         jsonelement.getAsJsonObject().addProperty("DataVersion", SharedConstants.getGameVersion().getWorldVersion());
 
+        // Yatopia start - replace whole logic
+        saveThread.submit(() -> {
+            try (OutputStreamWriter writer = new OutputStreamWriter(new FileOutputStream(f), Charsets.UTF_8.newEncoder())) {
+                AdvancementDataPlayer.b.toJson(jsonelement, writer);
+            } catch (Throwable e) {
+                AdvancementDataPlayer.LOGGER.error("Couldn't save player advancements to {}", this.f, e);
+            }
+        });
+        /*
         try {
             FileOutputStream fileoutputstream = new FileOutputStream(this.f);
             Throwable throwable = null;
@@ -275,6 +285,8 @@ public class AdvancementDataPlayer {
         } catch (IOException ioexception) {
             AdvancementDataPlayer.LOGGER.error("Couldn't save player advancements to {}", this.f, ioexception);
         }
+         */
+        // Yatopia end
 
     }
 
diff --git a/src/main/java/net/minecraft/server/EntityHuman.java b/src/main/java/net/minecraft/server/EntityHuman.java
index e411f6280d07e3f6a953e1fcaffaaa49f85dc7cc..68c033e2b8542c5656d5b9534a4953a370999097 100644
--- a/src/main/java/net/minecraft/server/EntityHuman.java
+++ b/src/main/java/net/minecraft/server/EntityHuman.java
@@ -700,11 +700,20 @@ public abstract class EntityHuman extends EntityLiving {
 
     }
 
+    // Yatopia start
+    private NBTTagList inventorySnapshot = null;
+    private NBTTagList enderchestSnapshot = null;
+    public void takeInventorySnapshot() {
+        inventorySnapshot = this.inventory.a(new NBTTagList());
+        enderchestSnapshot = this.enderChest.g();
+    }
+    // Yatopia end
+
     @Override
     public void saveData(NBTTagCompound nbttagcompound) {
         super.saveData(nbttagcompound);
         nbttagcompound.setInt("DataVersion", SharedConstants.getGameVersion().getWorldVersion());
-        nbttagcompound.set("Inventory", this.inventory.a(new NBTTagList()));
+        nbttagcompound.set("Inventory", inventorySnapshot != null ? inventorySnapshot : this.inventory.a(new NBTTagList())); inventorySnapshot = null;// Yatopia
         nbttagcompound.setInt("SelectedItemSlot", this.inventory.itemInHandIndex);
         nbttagcompound.setShort("SleepTimer", (short) this.sleepTicks);
         nbttagcompound.setFloat("XpP", this.exp);
@@ -714,7 +723,7 @@ public abstract class EntityHuman extends EntityLiving {
         nbttagcompound.setInt("Score", this.getScore());
         this.foodData.b(nbttagcompound);
         this.abilities.a(nbttagcompound);
-        nbttagcompound.set("EnderItems", this.enderChest.g());
+        nbttagcompound.set("EnderItems", enderchestSnapshot != null ? enderchestSnapshot : this.enderChest.g()); enderchestSnapshot = null;// Yatopia
         if (!this.getShoulderEntityLeft().isEmpty()) {
             nbttagcompound.set("ShoulderEntityLeft", this.getShoulderEntityLeft());
         }
diff --git a/src/main/java/net/minecraft/server/PlayerList.java b/src/main/java/net/minecraft/server/PlayerList.java
index 33c40bf94e74d1ee268fb8ffb809f45b31621df9..0b6c62ff0ac75c7ec11de9f8e9745b07ee12664e 100644
--- a/src/main/java/net/minecraft/server/PlayerList.java
+++ b/src/main/java/net/minecraft/server/PlayerList.java
@@ -1260,6 +1260,28 @@ public abstract class PlayerList {
             if (team != null) scoreboard.removeTeam(team);
         }
         // Paper end
+
+        // Yatopia start - make sure all saves are done
+        try {
+            playerFileData.saveThread.shutdown();
+			boolean done = playerFileData.saveThread.awaitTermination(60, java.util.concurrent.TimeUnit.SECONDS);
+			if (!done) {
+				LOGGER.error("Players did not save completly!");
+			}
+			ServerStatisticManager.saveThread.shutdown();
+			done = ServerStatisticManager.saveThread.awaitTermination(60, java.util.concurrent.TimeUnit.SECONDS);
+			if (!done) {
+				LOGGER.error("Stats did not save completly!");
+			}
+			AdvancementDataPlayer.saveThread.shutdown();
+			done = AdvancementDataPlayer.saveThread.awaitTermination(60, java.util.concurrent.TimeUnit.SECONDS);
+			if (!done) {
+				LOGGER.error("Advancements did not save completly!");
+			}
+		} catch (InterruptedException e) {
+			e.printStackTrace();
+		}
+        // Yatopia end
     }
     // Paper end
 
@@ -1297,13 +1319,13 @@ public abstract class PlayerList {
             File file = this.server.a(SavedFile.STATS).toFile();
             File file1 = new File(file, uuid + ".json");
 
-            if (!file1.exists()) {
+            /*if (!file1.exists()) { // Yatopia dont check for old stats files with sync File IO
                 File file2 = new File(file, displayName + ".json"); // CraftBukkit
 
                 if (file2.exists() && file2.isFile()) {
                     file2.renameTo(file1);
                 }
-            }
+            }*/
 
             serverstatisticmanager = new ServerStatisticManager(this.server, file1);
             // this.o.put(uuid, serverstatisticmanager); // CraftBukkit
diff --git a/src/main/java/net/minecraft/server/ServerStatisticManager.java b/src/main/java/net/minecraft/server/ServerStatisticManager.java
index 18725272f920eafb655de8b923a8a85dbdc704cf..a734cb47df9cffaee52a5c798bb3324212a004ba 100644
--- a/src/main/java/net/minecraft/server/ServerStatisticManager.java
+++ b/src/main/java/net/minecraft/server/ServerStatisticManager.java
@@ -20,6 +20,7 @@ import java.util.Map;
 import java.util.Map.Entry;
 import java.util.Optional;
 import java.util.Set;
+
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
@@ -30,6 +31,7 @@ public class ServerStatisticManager extends StatisticManager {
     private final File d;
     private final Set<Statistic<?>> e = Sets.newHashSet();
     private int f = -300;
+    public static java.util.concurrent.ExecutorService saveThread = java.util.concurrent.Executors.newSingleThreadExecutor(); // Yatopia
 
     public ServerStatisticManager(MinecraftServer minecraftserver, File file) {
         this.c = minecraftserver;
@@ -41,6 +43,7 @@ public class ServerStatisticManager extends StatisticManager {
             this.a.put( wrapper, entry.getValue().intValue() );
         }
         // Spigot end
+        saveThread.submit(() -> { // Yatopia
         if (file.isFile()) {
             try {
                 this.a(minecraftserver.getDataFixer(), org.apache.commons.io.FileUtils.readFileToString(file));
@@ -50,17 +53,21 @@ public class ServerStatisticManager extends StatisticManager {
                 ServerStatisticManager.LOGGER.error("Couldn't parse statistics file {}", file, jsonparseexception);
             }
         }
-
+        }); // Yatopia
     }
 
     public void save() {
         if ( org.spigotmc.SpigotConfig.disableStatSaving ) return; // Spigot
+        // Yatopia start
+        String str = this.b();
+        saveThread.submit(() -> {
         try {
-            org.apache.commons.io.FileUtils.writeStringToFile(this.d, this.b());
+            org.apache.commons.io.FileUtils.writeStringToFile(this.d, str);
         } catch (IOException ioexception) {
             ServerStatisticManager.LOGGER.error("Couldn't save stats", ioexception);
         }
-
+        });
+        // Yatopia end
     }
 
     @Override
@@ -111,7 +118,7 @@ public class ServerStatisticManager extends StatisticManager {
 
                                         if (nbttagcompound2.hasKeyOfType(s2, 99)) {
                                             SystemUtils.a(this.a(statisticwrapper, s2), (statistic) -> {
-                                                this.a.put(statistic, nbttagcompound2.getInt(s2));
+                                                this.a.put(statistic, nbttagcompound2.getInt(s2) + this.a.getOrDefault(statistic, 0)); // Yatopia fix async load
                                             }, () -> {
                                                 ServerStatisticManager.LOGGER.warn("Invalid statistic in {}: Don't know what {} is", this.d, s2);
                                             });
diff --git a/src/main/java/net/minecraft/server/WorldNBTStorage.java b/src/main/java/net/minecraft/server/WorldNBTStorage.java
index be060f602f53344cf51c22eafacb4dc9219c8856..12ffee96cfbd135f107b1dff7a058177e2649964 100644
--- a/src/main/java/net/minecraft/server/WorldNBTStorage.java
+++ b/src/main/java/net/minecraft/server/WorldNBTStorage.java
@@ -9,8 +9,6 @@ import java.io.OutputStream;
 import javax.annotation.Nullable;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
-
-// CraftBukkit start
 import org.bukkit.craftbukkit.entity.CraftPlayer;
 // CraftBukkit end
 
@@ -19,6 +17,7 @@ public class WorldNBTStorage {
     private static final Logger LOGGER = LogManager.getLogger();
     private final File playerDir;
     protected final DataFixer a;
+    public java.util.concurrent.ExecutorService saveThread = java.util.concurrent.Executors.newSingleThreadExecutor(); // Yatopia
 
     public WorldNBTStorage(Convertable.ConversionSession convertable_conversionsession, DataFixer datafixer) {
         this.a = datafixer;
@@ -28,11 +27,13 @@ public class WorldNBTStorage {
 
     public void save(EntityHuman entityhuman) {
         if(!com.destroystokyo.paper.PaperConfig.savePlayerData) return; // Paper - Make player data saving configurable
+        entityhuman.takeInventorySnapshot(); // Yatopia
+        saveThread.submit(() -> { // Yatopia
         try {
             NBTTagCompound nbttagcompound = entityhuman.save(new NBTTagCompound());
             File file = File.createTempFile(entityhuman.getUniqueIDString() + "-", ".dat", this.playerDir);
 
-            NBTCompressedStreamTools.a(nbttagcompound, (OutputStream) (new FileOutputStream(file)));
+            NBTCompressedStreamTools.a(nbttagcompound, new FileOutputStream(file));
             File file1 = new File(this.playerDir, entityhuman.getUniqueIDString() + ".dat");
             File file2 = new File(this.playerDir, entityhuman.getUniqueIDString() + ".dat_old");
 
@@ -40,6 +41,7 @@ public class WorldNBTStorage {
         } catch (Exception exception) {
             WorldNBTStorage.LOGGER.error("Failed to save player data for {}", entityhuman.getName(), exception); // Paper
         }
+        }); // Yatopia
 
     }
 
