From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: tr7zw <tr7zw@live.de>
Date: Tue, 21 Apr 2020 16:14:10 +0200
Subject: [PATCH] Modify default configs


diff --git a/src/main/java/com/destroystokyo/paper/PaperConfig.java b/src/main/java/com/destroystokyo/paper/PaperConfig.java
<<<<<<< HEAD:patches/server/0010-Modify-default-configs.patch
index 741c06a27e664211a7023a6369f8e69cbb41a321..f48fb8c3e637d0a7259bc92e5fb090c346750e47 100644
=======
index 56e4359ba32339e1bef58061585ff3e12e4215f3..8b607a76ab5cb6beadff3dc5e6832a394982063e 100644
>>>>>>> 1.16.1:patches/server/0015-Modify-default-configs.patch
--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
@@ -201,7 +201,7 @@ public class PaperConfig {
 
     public static String timingsServerName;
     private static void timings() {
-        boolean timings = getBoolean("timings.enabled", true);
+        boolean timings = getBoolean("timings.enabled", false); // YAPFA don't profile by default
         boolean verboseTimings = getBoolean("timings.verbose", true);
         TimingsManager.privacy = getBoolean("timings.server-name-privacy", false);
         TimingsManager.hiddenConfigs = getList("timings.hidden-config-entries", Lists.newArrayList("database", "settings.bungeecord-addresses"));
diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index e471e764935e2a89560de56959a782b02e5e8fe1..b0b497c8cdddb155722c89e69051b8c7519cd4d4 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -600,7 +600,7 @@ public class PaperWorldConfig {
     }
 
     public boolean cooldownHopperWhenFull = true;
-    public boolean disableHopperMoveEvents = false;
+    public boolean disableHopperMoveEvents = true; // YAPFA disable by default
     private void hopperOptimizations() {
         cooldownHopperWhenFull = getBoolean("hopper.cooldown-when-full", cooldownHopperWhenFull);
         log("Cooldown Hoppers when Full: " + (cooldownHopperWhenFull ? "enabled" : "disabled"));
