From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: tr7zw <tr7zw@live.de>
Date: Fri, 31 Jul 2020 21:48:14 -0500
Subject: [PATCH] Item stuck sleep config


diff --git a/src/main/java/net/minecraft/world/entity/item/EntityItem.java b/src/main/java/net/minecraft/world/entity/item/EntityItem.java
index cb5c4379bfd16a74debaf23f840dbf1e5535ad7f..392f51e435021352cf14554361fc6425b70d06e6 100644
--- a/src/main/java/net/minecraft/world/entity/item/EntityItem.java
+++ b/src/main/java/net/minecraft/world/entity/item/EntityItem.java
@@ -114,7 +114,7 @@ public class EntityItem extends Entity {
                 if (gg.airplane.AirplaneConfig.simplerItemCollision) {
                     net.minecraft.world.level.block.state.IBlockData block = this.world.getTypeIfLoaded(this.getChunkCoordinates());
                     this.noclip = block != null && block.isOpaque();
-                } else
+                } else if (!this.onGround || this.noclip || this.ticksLived % org.yatopiamc.yatopia.server.YatopiaConfig.itemStuckSleepTicks == 0)
                     // Airplane end
                 this.noclip = !this.world.getCubes(this);
                 if (this.noclip) {
diff --git a/src/main/java/org/yatopiamc/yatopia/server/YatopiaConfig.java b/src/main/java/org/yatopiamc/yatopia/server/YatopiaConfig.java
index f5a43e607f9c13e07a0f5e7e139eed93f159185b..4722a170519dfebc7b41ce886563a3b80c8e111b 100644
--- a/src/main/java/org/yatopiamc/yatopia/server/YatopiaConfig.java
+++ b/src/main/java/org/yatopiamc/yatopia/server/YatopiaConfig.java
@@ -189,4 +189,9 @@ public class YatopiaConfig {
     private static void brandName() {
         brandName = getString("brand-name", brandName);
     }
+
+    public static int itemStuckSleepTicks = 1;
+    private static void itemStuckSleepTicks() {
+        itemStuckSleepTicks = getInt("settings.itemStuckSleepTicks", 1);
+    }
 }
