From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: tr7zw <tr7zw@live.de>
Date: Sat, 25 Jul 2020 17:16:18 +0200
Subject: [PATCH] Global Eula file


diff --git a/src/main/java/net/minecraft/server/EULA.java b/src/main/java/net/minecraft/server/EULA.java
index 3bc5cd1e53dd7c94b948e7f57f0dc8e073e349b0..3782e6ac4a5548877ece723f967f879459853d3a 100644
--- a/src/main/java/net/minecraft/server/EULA.java
+++ b/src/main/java/net/minecraft/server/EULA.java
@@ -17,12 +17,25 @@ public class EULA {
 
     public EULA(Path path) {
         this.b = path;
-        this.c = SharedConstants.d || this.b();
+        this.c = SharedConstants.d || globalEula() || this.b(this.b); // Yatopia
     }
 
-    private boolean b() {
+    // Yatopia start
+    private boolean globalEula() {
+	    java.io.File globalEula = new java.io.File(System.getProperty("user.home"), "eula.txt");
+
+	    if (globalEula.exists()) {
+		    return b(globalEula.toPath());
+	    } else {
+		    System.out.println("No global eula found at " + globalEula.getAbsolutePath());
+	    }
+	    return false;
+    }
+    // Yatopia end
+
+    private boolean b(java.nio.file.Path path) { // Yatopia
         try {
-            InputStream inputstream = Files.newInputStream(this.b);
+            InputStream inputstream = Files.newInputStream(path);
             Throwable throwable = null;
 
             boolean flag;
