From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: tr7zw <tr7zw@live.de>
Date: Tue, 4 Aug 2020 22:08:09 +0200
Subject: [PATCH] Fix lead fall dmg config


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index d52ac77516cbf62293afbbfdd96f5a4861a3025f..da6c87202d5d9c4b367bebd28e94484506604184 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -1426,6 +1426,8 @@ public abstract class Entity implements INamableTileEntity, ICommandListener, ne
             this.fallDistance = 0.0F;
         } else if (d0 < 0.0D) {
             this.fallDistance = (float) ((double) this.fallDistance - d0);
+        } else if (d0 > 0.0D && org.yatopiamc.yatopia.server.YatopiaConfig.fixFallDistance) { // Yatopia
+            this.fallDistance = 0.0F; // Yatopia
         }
 
     }
diff --git a/src/main/java/org/yatopiamc/yatopia/server/YatopiaConfig.java b/src/main/java/org/yatopiamc/yatopia/server/YatopiaConfig.java
index da7e3d5179a4328d4a74256caa697f5d45725e59..43c6d26dcf2d8875bc5449478acc29a25a88f2b8 100644
--- a/src/main/java/org/yatopiamc/yatopia/server/YatopiaConfig.java
+++ b/src/main/java/org/yatopiamc/yatopia/server/YatopiaConfig.java
@@ -231,4 +231,9 @@ public class YatopiaConfig {
         simplerVillagerBehavior = getBoolean("settings.villager.simplerVillagerBehavior", false);
         villagersHideAtNight = getBoolean("settings.villager.villagersHideAtNight", false);
     }
+
+    public static boolean fixFallDistance = false;
+    private static void fixFallDistance() {
+        fixFallDistance = getBoolean("settings.fixFallDistance", false);
+    }
 }
