From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: tr7zw <tr7zw@live.de>
Date: Tue, 4 Aug 2020 22:08:09 +0200
Subject: [PATCH] Fix lead fall dmg config


diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index ae89d7074c5c9fd499bb1492a835362a005d1cfd..a9b89e54184b08467426f37e81d9510af227b47f 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -1307,6 +1307,8 @@ public abstract class Entity implements INamableTileEntity, ICommandListener, Ke
             this.fallDistance = 0.0F;
         } else if (d0 < 0.0D) {
             this.fallDistance = (float) ((double) this.fallDistance - d0);
+        } else if (d0 > 0.0D && org.yatopiamc.yatopia.server.YatopiaConfig.fixFallDistance) { // Yatopia
+            this.fallDistance = 0.0F; // Yatopia
         }
 
     }
diff --git a/src/main/java/org/yatopiamc/yatopia/server/YatopiaConfig.java b/src/main/java/org/yatopiamc/yatopia/server/YatopiaConfig.java
index 0f3803757a81c7ecaad48637ede8a3dc95b93f13..003a024da8ab877895244ff9e4e4ff62288622ff 100644
--- a/src/main/java/org/yatopiamc/yatopia/server/YatopiaConfig.java
+++ b/src/main/java/org/yatopiamc/yatopia/server/YatopiaConfig.java
@@ -206,4 +206,9 @@ public class YatopiaConfig {
         simplerVillagerBehavior = getBoolean("settings.villager.simplerVillagerBehavior", false);
         villagersHideAtNight = getBoolean("settings.villager.villagersHideAtNight", false);
     }
+
+    public static boolean fixFallDistance = false;
+    private static void fixFallDistance() {
+        fixFallDistance = getBoolean("settings.fixFallDistance", false);
+    }
 }
