From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ivan Pekov <ivan@mrivanplays.com>
Date: Tue, 18 Aug 2020 07:53:25 +0300
Subject: [PATCH] Fix MC-197616


diff --git a/src/main/java/net/minecraft/server/PacketPlayOutMapChunk.java b/src/main/java/net/minecraft/server/PacketPlayOutMapChunk.java
index 72fdbf1534b65284ac8020dcc15fe1512766d087..e4f3157265df8208e6ef4ecf630e073f69cd9a38 100644
--- a/src/main/java/net/minecraft/server/PacketPlayOutMapChunk.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutMapChunk.java
@@ -175,6 +175,7 @@ public class PacketPlayOutMapChunk implements Packet<PacketListenerPlayOut> {
         packetdataserializer.d(this.c);
         packetdataserializer.a(this.d);
         if (this.e != null) {
+            for (int i = 0; i < e.length; i++) { if (e[i] < 0) e[i] = 0; } // Yatopia
             packetdataserializer.a(this.e);
         }
 
