From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ivan Pekov <ivan@mrivanplays.com>
Date: Thu, 10 Sep 2020 21:02:26 +0300
Subject: [PATCH] Fix LightEngineThreaded memory leak


diff --git a/src/main/java/net/minecraft/server/LightEngineThreaded.java b/src/main/java/net/minecraft/server/LightEngineThreaded.java
index d4902ed0d12d9697402ca60bb8a298f753ccf527..9f5142881d1b46a800cffddcf2417500d549aef4 100644
--- a/src/main/java/net/minecraft/server/LightEngineThreaded.java
+++ b/src/main/java/net/minecraft/server/LightEngineThreaded.java
@@ -109,6 +109,8 @@ public class LightEngineThreaded extends LightEngine implements AutoCloseable {
             }
         }
 
+        public final void remove(long chunkId) { pendingTasks.removeIf(task -> task.chunkId == chunkId); queueUpdate(); } // Yatopia
+
         public final boolean isEmpty() {
             return this.size == 0 && this.pendingTasks.isEmpty();
         }
diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index debc82608521d0d44da3e5a4c4dd8525b7e48109..6381cf60244737f37f88d7daa01f25138a58c3ac 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -1790,6 +1790,7 @@ public class WorldServer extends World implements GeneratorAccessSeed {
         }
         // Paper end
 
+        chunkProvider.getLightEngine().queue.remove(chunk.getPos().pair()); // Yatopia - unregister from light queue
     }
 
     public void unregisterEntity(Entity entity) {
