From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Simon L <slooij1@gmail.com>
Date: Sat, 29 Aug 2020 19:19:47 +0200
Subject: [PATCH] Fix: Lithium oversized blocks counter not counting blocks
 correctly


diff --git a/src/main/java/net/minecraft/server/ChunkSection.java b/src/main/java/net/minecraft/server/ChunkSection.java
index c2d584759f290a4766ee077d6d05e659bd2aecf5..478af91f52272424da61d626534298290d6385f5 100644
--- a/src/main/java/net/minecraft/server/ChunkSection.java
+++ b/src/main/java/net/minecraft/server/ChunkSection.java
@@ -68,6 +68,9 @@ public class ChunkSection implements me.jellysquid.mods.lithium.common.entity.mo
 
         if (!iblockdata1.isAir()) {
             --this.nonEmptyBlockCount;
+            if(iblockdata1.shapeExceedsCube()){ //Yatopia - lithium oversized blocks count
+                --this.oversizedBlockCount;
+            } // Yatopia end
             if (iblockdata1.isTicking()) {
                 --this.tickingBlockCount;
                 // Paper start
@@ -82,6 +85,9 @@ public class ChunkSection implements me.jellysquid.mods.lithium.common.entity.mo
 
         if (!iblockdata.isAir()) {
             ++this.nonEmptyBlockCount;
+            if(iblockdata.shapeExceedsCube()){ //Yatopia - lithium oversized blocks count
+                ++this.oversizedBlockCount;
+            } // Yatopia end
             if (iblockdata.isTicking()) {
                 ++this.tickingBlockCount;
                 // Paper start
