From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mariell Hoversholm <proximyst@proximyst.com>
Date: Mon, 28 Sep 2020 16:25:34 +0300
Subject: [PATCH] Fix merging spawning values


diff --git a/src/main/java/net/minecraft/server/SpawnerCreature.java b/src/main/java/net/minecraft/server/SpawnerCreature.java
index 9108321741e9d3f896c43be8f204735b1b73127e..b16ca40d49a0cc8f4fb3b1de1ce695b367040ad4 100644
--- a/src/main/java/net/minecraft/server/SpawnerCreature.java
+++ b/src/main/java/net/minecraft/server/SpawnerCreature.java
@@ -157,9 +157,12 @@ public final class SpawnerCreature {
                     spawnercreature_d.a(entityinsentient, ichunkaccess);
                 },
                 difference, worldserver.paperConfig.perPlayerMobSpawns ? worldserver.getChunkProvider().playerChunkMap::updatePlayerMobTypeMap : null);
-                spawnercreature_d.getEntityCountsByType().mergeInt(enumcreaturetype, spawnCount, (keyInMap, valueInMap) -> {
-                    return Integer.valueOf(spawnCount + valueInMap.intValue());
-                });
+                // Yatopia start - fix this
+                //spawnercreature_d.getEntityCountsByType().mergeInt(enumcreaturetype, spawnCount, (keyInMap, valueInMap) -> {
+                //    return Integer.valueOf(spawnCount + valueInMap.intValue());
+                //});
+                spawnercreature_d.getEntityCountsByType().mergeInt(enumcreaturetype, spawnCount, Integer::sum);
+                // Yatopia end
                 // Paper end - per player mob spawning
             }
         }
