From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ivan Pekov <ivan@mrivanplays.com>
Date: Mon, 28 Sep 2020 16:23:11 +0300
Subject: [PATCH] Respect rotation when respawning


diff --git a/src/main/java/net/minecraft/server/PlayerList.java b/src/main/java/net/minecraft/server/PlayerList.java
index 3df9a91c0474df8a2884f4a4ae7d016bd31600e9..171aa373e1ecd30211ec04810441c8f9cc1039f7 100644
--- a/src/main/java/net/minecraft/server/PlayerList.java
+++ b/src/main/java/net/minecraft/server/PlayerList.java
@@ -815,7 +815,7 @@ public abstract class PlayerList {
                     entityplayer1.setRespawnPosition(worldserver1.getDimensionKey(), blockposition, f, flag1, false);
                     flag2 = !flag && flag3;
                     isBedSpawn = true;
-                    location = new Location(worldserver1.getWorld(), vec3d.x, vec3d.y, vec3d.z);
+                    location = new Location(worldserver1.getWorld(), vec3d.x, vec3d.y, vec3d.z, f1, 0.0F); // Yatopia - respect rotation
                 } else if (blockposition != null) {
                     entityplayer1.playerConnection.sendPacket(new PacketPlayOutGameStateChange(PacketPlayOutGameStateChange.a, 0.0F));
                 }
