From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bud Gidiere <sgidiere@gmail.com>
Date: Sun, 25 Oct 2020 12:24:19 -0500
Subject: [PATCH] Infinity No Arrows


diff --git a/src/main/java/net/minecraft/server/EntityHuman.java b/src/main/java/net/minecraft/server/EntityHuman.java
index 21074d369b9f7c7f101ba85a9d350627b98217d2..e35087458dea5ac1736ec2d230d8c23b84800774 100644
--- a/src/main/java/net/minecraft/server/EntityHuman.java
+++ b/src/main/java/net/minecraft/server/EntityHuman.java
@@ -2137,7 +2137,7 @@ public abstract class EntityHuman extends EntityLiving {
                     }
                 }
 
-                return this.abilities.canInstantlyBuild ? new ItemStack(Items.ARROW) : ItemStack.b;
+                return this.abilities.canInstantlyBuild || (world.purpurConfig.infinityWorksWithNoArrows && EnchantmentManager.getEnchantmentLevel(Enchantments.ARROW_INFINITE, itemstack) > 0) ? new ItemStack(Items.ARROW) : ItemStack.b; // Yatopia
             }
         }
     }
diff --git a/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java b/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java
index 0432d0320aca70750759935c0f2bd2ef737a596f..fb6a5f3b60819c33a3fed43f2e80558fc5d54d5e 100644
--- a/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java
+++ b/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java
@@ -177,10 +177,12 @@ public class PurpurWorldConfig {
     public boolean infinityWorksWithNormalArrows = true;
     public boolean infinityWorksWithSpectralArrows = false;
     public boolean infinityWorksWithTippedArrows = false;
+    public boolean infinityWorksWithNoArrows = false;
     private void infinityArrowsSettings() {
         infinityWorksWithNormalArrows = getBoolean("gameplay-mechanics.infinity-bow.normal-arrows", infinityWorksWithNormalArrows);
         infinityWorksWithSpectralArrows = getBoolean("gameplay-mechanics.infinity-bow.spectral-arrows", infinityWorksWithSpectralArrows);
         infinityWorksWithTippedArrows = getBoolean("gameplay-mechanics.infinity-bow.tipped-arrows", infinityWorksWithTippedArrows);
+        infinityWorksWithNoArrows = getBoolean("gameplay-mechanics.infinity-bow.no-arrows", infinityWorksWithNoArrows);
     }
 
     public boolean signAllowColors = false;
