From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Simon Gardling <titaniumtown@gmail.com>
Date: Wed, 16 Dec 2020 15:23:46 -0500
Subject: [PATCH] Configurable max world size


diff --git a/src/main/java/net/minecraft/server/BaseBlockPosition.java b/src/main/java/net/minecraft/server/BaseBlockPosition.java
index e811295b4d6afcd920f60e0ce5440e43300d9085..172f38765533de745f6d601c719e914c43454e62 100644
--- a/src/main/java/net/minecraft/server/BaseBlockPosition.java
+++ b/src/main/java/net/minecraft/server/BaseBlockPosition.java
@@ -22,7 +22,7 @@ public class BaseBlockPosition implements Comparable<BaseBlockPosition> {
 
     // Paper start
     public boolean isValidLocation() {
-        return getX() >= -30000000 && getZ() >= -30000000 && getX() < 30000000 && getZ() < 30000000 && getY() >= 0 && getY() < 256;
+        return getX() >= -org.yatopiamc.yatopia.server.YatopiaConfig.biggestInt && getZ() >= -org.yatopiamc.yatopia.server.YatopiaConfig.biggestInt && getX() < org.yatopiamc.yatopia.server.YatopiaConfig.biggestInt && getZ() < org.yatopiamc.yatopia.server.YatopiaConfig.biggestInt && getY() >= 0 && getY() < 256; // Yatopia - Remove world border limitations
     }
     public boolean isInvalidYLocation() {
         return b < 0 || b >= 256;
diff --git a/src/main/java/net/minecraft/server/CommandForceload.java b/src/main/java/net/minecraft/server/CommandForceload.java
index 57e9e737f1d3c82eb81bfd2a6786ec98478d5615..a94217dd65bab5f4e612faecff6262cc5732aac7 100644
--- a/src/main/java/net/minecraft/server/CommandForceload.java
+++ b/src/main/java/net/minecraft/server/CommandForceload.java
@@ -8,6 +8,7 @@ import com.mojang.brigadier.exceptions.CommandSyntaxException;
 import com.mojang.brigadier.exceptions.Dynamic2CommandExceptionType;
 import com.mojang.brigadier.exceptions.SimpleCommandExceptionType;
 import it.unimi.dsi.fastutil.longs.LongSet;
+import java.util.function.Consumer;
 
 public class CommandForceload {
 
@@ -80,7 +81,7 @@ public class CommandForceload {
         ResourceKey<World> resourcekey = worldserver.getDimensionKey();
         LongSet longset = worldserver.getForceLoadedChunks();
 
-        longset.forEach((i) -> {
+        longset.forEach((Consumer<? super Long>) (i) -> {  // Yatopia - Remove world border limitations
             worldserver.setForceLoaded(ChunkCoordIntPair.getX(i), ChunkCoordIntPair.getZ(i), false);
         });
         commandlistenerwrapper.sendMessage(new ChatMessage("commands.forceload.removed.all", new Object[]{resourcekey.a()}), true);
@@ -93,7 +94,7 @@ public class CommandForceload {
         int k = Math.max(blockposition2d.a, blockposition2d1.a);
         int l = Math.max(blockposition2d.b, blockposition2d1.b);
 
-        if (i >= -30000000 && j >= -30000000 && k < 30000000 && l < 30000000) {
+        if (i >= -org.yatopiamc.yatopia.server.YatopiaConfig.biggestInt && j >= -org.yatopiamc.yatopia.server.YatopiaConfig.biggestInt && k < org.yatopiamc.yatopia.server.YatopiaConfig.biggestInt && l < org.yatopiamc.yatopia.server.YatopiaConfig.biggestInt) {  // Yatopia - Remove world border limitations
             int i1 = i >> 4;
             int j1 = j >> 4;
             int k1 = k >> 4;
diff --git a/src/main/java/net/minecraft/server/CommandTeleport.java b/src/main/java/net/minecraft/server/CommandTeleport.java
index 57a23dd7a1196402ddc58e7491ac9ada97f402a4..ed20d13a9c6caf8054f2ac0d53933ae94b8c1353 100644
--- a/src/main/java/net/minecraft/server/CommandTeleport.java
+++ b/src/main/java/net/minecraft/server/CommandTeleport.java
@@ -121,7 +121,7 @@ public class CommandTeleport {
     private static void a(CommandListenerWrapper commandlistenerwrapper, Entity entity, WorldServer worldserver, double d0, double d1, double d2, Set<PacketPlayOutPosition.EnumPlayerTeleportFlags> set, float f, float f1, @Nullable CommandTeleport.a commandteleport_a) throws CommandSyntaxException {
         BlockPosition blockposition = new BlockPosition(d0, d1, d2);
         // Paper start - Don't allow teleport command to invalid locations
-        if (d0 <= -30000000 || d2 <= -30000000 || d0 > 30000000 || d2 > 30000000 || d1 > 30000000 || d1 <= -30000000) { // Copy/pasta from BaseBlockPosition#isValidLocation
+        if (d0 <= -org.yatopiamc.yatopia.server.YatopiaConfig.biggestInt || d2 <= -org.yatopiamc.yatopia.server.YatopiaConfig.biggestInt || d0 > org.yatopiamc.yatopia.server.YatopiaConfig.biggestInt || d2 > org.yatopiamc.yatopia.server.YatopiaConfig.biggestInt || d1 > org.yatopiamc.yatopia.server.YatopiaConfig.biggestInt || d1 <= -org.yatopiamc.yatopia.server.YatopiaConfig.biggestInt) { // Copy/pasta from BaseBlockPosition#isValidLocation  // Yatopia - Remove world border limitations
             org.bukkit.Bukkit.getLogger().warning("Refused to teleport " + entity.getName() + " to " + d0 + ", " + d1 + ", " + d2);
             return;
         }
diff --git a/src/main/java/net/minecraft/server/CommandWorldBorder.java b/src/main/java/net/minecraft/server/CommandWorldBorder.java
index 11ad785b2408ba94f9d80c4af3a81e0c0e7e9a64..febaa6246c34e536961e07cc62063024246dc6fc 100644
--- a/src/main/java/net/minecraft/server/CommandWorldBorder.java
+++ b/src/main/java/net/minecraft/server/CommandWorldBorder.java
@@ -23,11 +23,11 @@ public class CommandWorldBorder {
     public static void a(com.mojang.brigadier.CommandDispatcher<CommandListenerWrapper> com_mojang_brigadier_commanddispatcher) {
         com_mojang_brigadier_commanddispatcher.register((LiteralArgumentBuilder) ((LiteralArgumentBuilder) ((LiteralArgumentBuilder) ((LiteralArgumentBuilder) ((LiteralArgumentBuilder) ((LiteralArgumentBuilder) ((LiteralArgumentBuilder) CommandDispatcher.a("worldborder").requires((commandlistenerwrapper) -> {
             return commandlistenerwrapper.hasPermission(2);
-        })).then(CommandDispatcher.a("add").then(((RequiredArgumentBuilder) CommandDispatcher.a("distance", (ArgumentType) FloatArgumentType.floatArg(-6.0E7F, 6.0E7F)).executes((commandcontext) -> {
+        })).then(CommandDispatcher.a("add").then(((RequiredArgumentBuilder) CommandDispatcher.a("distance", (ArgumentType) FloatArgumentType.floatArg(-org.yatopiamc.yatopia.server.YatopiaConfig.biggestFloat, org.yatopiamc.yatopia.server.YatopiaConfig.biggestFloat)).executes((commandcontext) -> {
             return a((CommandListenerWrapper) commandcontext.getSource(), ((CommandListenerWrapper) commandcontext.getSource()).getWorld().getWorldBorder().getSize() + (double) FloatArgumentType.getFloat(commandcontext, "distance"), 0L);
         })).then(CommandDispatcher.a("time", (ArgumentType) IntegerArgumentType.integer(0)).executes((commandcontext) -> {
             return a((CommandListenerWrapper) commandcontext.getSource(), ((CommandListenerWrapper) commandcontext.getSource()).getWorld().getWorldBorder().getSize() + (double) FloatArgumentType.getFloat(commandcontext, "distance"), ((CommandListenerWrapper) commandcontext.getSource()).getWorld().getWorldBorder().j() + (long) IntegerArgumentType.getInteger(commandcontext, "time") * 1000L);
-        }))))).then(CommandDispatcher.a("set").then(((RequiredArgumentBuilder) CommandDispatcher.a("distance", (ArgumentType) FloatArgumentType.floatArg(-6.0E7F, 6.0E7F)).executes((commandcontext) -> {
+        }))))).then(CommandDispatcher.a("set").then(((RequiredArgumentBuilder) CommandDispatcher.a("distance", (ArgumentType) FloatArgumentType.floatArg(-org.yatopiamc.yatopia.server.YatopiaConfig.biggestFloat, org.yatopiamc.yatopia.server.YatopiaConfig.biggestFloat)).executes((commandcontext) -> {
             return a((CommandListenerWrapper) commandcontext.getSource(), (double) FloatArgumentType.getFloat(commandcontext, "distance"), 0L);
         })).then(CommandDispatcher.a("time", (ArgumentType) IntegerArgumentType.integer(0)).executes((commandcontext) -> {
             return a((CommandListenerWrapper) commandcontext.getSource(), (double) FloatArgumentType.getFloat(commandcontext, "distance"), (long) IntegerArgumentType.getInteger(commandcontext, "time") * 1000L);
@@ -121,7 +121,7 @@ public class CommandWorldBorder {
             throw CommandWorldBorder.b.create();
         } else if (d0 < 1.0D) {
             throw CommandWorldBorder.c.create();
-        } else if (d0 > 6.0E7D) {
+        } else if (d0 > org.yatopiamc.yatopia.server.YatopiaConfig.WborderSize) {  // Yatopia - Remove world border limitations
             throw CommandWorldBorder.d.create();
         } else {
             if (i > 0L) {
diff --git a/src/main/java/net/minecraft/server/DedicatedServer.java b/src/main/java/net/minecraft/server/DedicatedServer.java
index bd43204760663363fcc53c1a8c175668809f96b0..60f3409acb905c1f9ccae3b2322092f6c4876560 100644
--- a/src/main/java/net/minecraft/server/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/DedicatedServer.java
@@ -562,7 +562,7 @@ public class DedicatedServer extends MinecraftServer implements IMinecraftServer
 
     @Override
     public int au() {
-        return this.getDedicatedServerProperties().maxWorldSize;
+        return org.yatopiamc.yatopia.server.YatopiaConfig.biggestInt; // Yatopia - Remove world border limitations
     }
 
     @Override
diff --git a/src/main/java/net/minecraft/server/DedicatedServerProperties.java b/src/main/java/net/minecraft/server/DedicatedServerProperties.java
index fe5937d890762ab37a90f189c4992af389d2c41d..a9f3a1dda2c8408787548c6ea200cefee09751b9 100644
--- a/src/main/java/net/minecraft/server/DedicatedServerProperties.java
+++ b/src/main/java/net/minecraft/server/DedicatedServerProperties.java
@@ -52,7 +52,6 @@ public class DedicatedServerProperties extends PropertyManager<DedicatedServerPr
     public final int networkCompressionThreshold;
     public final boolean broadcastRconToOps;
     public final boolean broadcastConsoleToOps;
-    public final int maxWorldSize;
     public final boolean syncChunkWrites;
     public final boolean enableJmxMonitoring;
     public final boolean enableStatus;
@@ -103,9 +102,6 @@ public class DedicatedServerProperties extends PropertyManager<DedicatedServerPr
         this.networkCompressionThreshold = this.getInt("network-compression-threshold", 256);
         this.broadcastRconToOps = this.getBoolean("broadcast-rcon-to-ops", true);
         this.broadcastConsoleToOps = this.getBoolean("broadcast-console-to-ops", true);
-        this.maxWorldSize = this.a("max-world-size", (integer) -> {
-            return MathHelper.clamp(integer, 1, 29999984);
-        }, 29999984);
         this.syncChunkWrites = this.getBoolean("sync-chunk-writes", true) && Boolean.getBoolean("Paper.enable-sync-chunk-writes"); // Paper - hide behind flag
         this.enableJmxMonitoring = this.getBoolean("enable-jmx-monitoring", false);
         this.enableStatus = this.getBoolean("enable-status", true);
diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index 7c84ba2169dd72177f8426d990da72b6f30b7e4f..b6f03a24346cfbed73acdc4a1be11e2ea6b30fb6 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -1553,8 +1553,8 @@ public abstract class Entity implements INamableTileEntity, ICommandListener, Ke
     }
 
     public void f(double d0, double d1, double d2) {
-        double d3 = MathHelper.a(d0, -3.0E7D, 3.0E7D);
-        double d4 = MathHelper.a(d2, -3.0E7D, 3.0E7D);
+        double d3 = MathHelper.a(d0, -org.yatopiamc.yatopia.server.YatopiaConfig.replace2, org.yatopiamc.yatopia.server.YatopiaConfig.replace2); // Yatopia - Remove world border limitations
+        double d4 = MathHelper.a(d2, -org.yatopiamc.yatopia.server.YatopiaConfig.replace2, org.yatopiamc.yatopia.server.YatopiaConfig.replace2); // Yatopia - Remove world border limitations
 
         this.lastX = d3;
         this.lastY = d1;
@@ -2927,10 +2927,12 @@ public abstract class Entity implements INamableTileEntity, ICommandListener, Ke
                 return null;
             } else {
                 WorldBorder worldborder = worldserver.getWorldBorder();
-                double d0 = Math.max(-2.9999872E7D, worldborder.e() + 16.0D);
-                double d1 = Math.max(-2.9999872E7D, worldborder.f() + 16.0D);
-                double d2 = Math.min(2.9999872E7D, worldborder.g() - 16.0D);
-                double d3 = Math.min(2.9999872E7D, worldborder.h() - 16.0D);
+                // Yatopia Start - Remove world border limitations
+                double d0 = Math.max(-org.yatopiamc.yatopia.server.YatopiaConfig.replace1, worldborder.e() + 16.0D);
+                double d1 = Math.max(-org.yatopiamc.yatopia.server.YatopiaConfig.replace1, worldborder.f() + 16.0D);
+                double d2 = Math.min(org.yatopiamc.yatopia.server.YatopiaConfig.replace1, worldborder.g() - 16.0D);
+                double d3 = Math.min(org.yatopiamc.yatopia.server.YatopiaConfig.replace1, worldborder.h() - 16.0D);
+                // Yatopia End - Remove world border limitations
                 double d4 = DimensionManager.a(this.world.getDimensionManager(), worldserver.getDimensionManager());
                 BlockPosition blockposition = new BlockPosition(MathHelper.a(this.locX() * d4, d0, d2), this.locY(), MathHelper.a(this.locZ() * d4, d1, d3));
                 // CraftBukkit start
diff --git a/src/main/java/net/minecraft/server/IWorldReader.java b/src/main/java/net/minecraft/server/IWorldReader.java
index eb7282c33dd4f0bb26b9ccafc42bd92e6fdb997e..6ffc96da4709de44c5190342ddbe8ce2642dcee6 100644
--- a/src/main/java/net/minecraft/server/IWorldReader.java
+++ b/src/main/java/net/minecraft/server/IWorldReader.java
@@ -139,7 +139,7 @@ public interface IWorldReader extends IBlockLightAccess, ICollisionAccess, Biome
     }
 
     default int c(BlockPosition blockposition, int i) {
-        return blockposition.getX() >= -30000000 && blockposition.getZ() >= -30000000 && blockposition.getX() < 30000000 && blockposition.getZ() < 30000000 ? this.getLightLevel(blockposition, i) : 15;
+        return blockposition.getX() >= -org.yatopiamc.yatopia.server.YatopiaConfig.replace3 && blockposition.getZ() >= -org.yatopiamc.yatopia.server.YatopiaConfig.replace3 && blockposition.getX() < org.yatopiamc.yatopia.server.YatopiaConfig.replace3 && blockposition.getZ() < org.yatopiamc.yatopia.server.YatopiaConfig.replace3 ? this.getLightLevel(blockposition, i) : 15; // Yatopia - Remove world border limitations
     }
 
     @Deprecated
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index c0be5153d117208edf69459728d57e11e26b8955..2803ad6285231a3711ffa3faa991560b13759e24 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1874,7 +1874,7 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
     }
 
     public int au() {
-        return 29999984;
+        return org.yatopiamc.yatopia.server.YatopiaConfig.biggestInt; // Yatopia - Remove world border limitations
     }
 
     @Override
diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 0364c923792acb30dc600db8ed4f05f8d5597e1d..ace0f40a19f7997e75cdd50d8873b19c7782ef1f 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -382,7 +382,7 @@ public class PlayerConnection implements PacketListenerPlayIn {
     }
 
     private static boolean b(PacketPlayInFlying packetplayinflying) {
-        return Doubles.isFinite(packetplayinflying.a(0.0D)) && Doubles.isFinite(packetplayinflying.b(0.0D)) && Doubles.isFinite(packetplayinflying.c(0.0D)) && Floats.isFinite(packetplayinflying.b(0.0F)) && Floats.isFinite(packetplayinflying.a(0.0F)) ? Math.abs(packetplayinflying.a(0.0D)) > 3.0E7D || Math.abs(packetplayinflying.b(0.0D)) > 3.0E7D || Math.abs(packetplayinflying.c(0.0D)) > 3.0E7D : true;
+        return Doubles.isFinite(packetplayinflying.a(0.0D)) && Doubles.isFinite(packetplayinflying.b(0.0D)) && Doubles.isFinite(packetplayinflying.c(0.0D)) && Floats.isFinite(packetplayinflying.b(0.0F)) && Floats.isFinite(packetplayinflying.a(0.0F)) ? Math.abs(packetplayinflying.a(0.0D)) > org.yatopiamc.yatopia.server.YatopiaConfig.replace2 || Math.abs(packetplayinflying.b(0.0D)) > org.yatopiamc.yatopia.server.YatopiaConfig.replace2 || Math.abs(packetplayinflying.c(0.0D)) > org.yatopiamc.yatopia.server.YatopiaConfig.replace2 : true; // Yatopia - Remove world border limitations
     }
 
     private static boolean b(PacketPlayInVehicleMove packetplayinvehiclemove) {
diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 55238930d8352ccedb963389461f119cd813aff9..8775ce10a3104f00ce9704a9a9f813b76586555f 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -460,7 +460,8 @@ public abstract class World implements GeneratorAccess, AutoCloseable {
     }
 
     private static boolean D(BlockPosition blockposition) {
-        return blockposition.getX() >= -30000000 && blockposition.getZ() >= -30000000 && blockposition.getX() < 30000000 && blockposition.getZ() < 30000000;
+        // return blockposition.getX() >= -30000000 && blockposition.getZ() >= -30000000 && blockposition.getX() < 30000000 && blockposition.getZ() < 30000000;
+        return blockposition.getX() >= -org.yatopiamc.yatopia.server.YatopiaConfig.biggestInt && blockposition.getZ() >= -org.yatopiamc.yatopia.server.YatopiaConfig.biggestInt && blockposition.getX() < org.yatopiamc.yatopia.server.YatopiaConfig.biggestInt && blockposition.getZ() < org.yatopiamc.yatopia.server.YatopiaConfig.biggestInt; // Yatopia - Remove world border limitations
     }
 
     private static boolean d(int i) {
@@ -849,7 +850,8 @@ public abstract class World implements GeneratorAccess, AutoCloseable {
     @Override public int a(HeightMap.Type heightmap_type, int i, int j) { // Paper - OBFHELPER
         int k;
 
-        if (i >= -30000000 && j >= -30000000 && i < 30000000 && j < 30000000) {
+        // if (i >= -30000000 && j >= -30000000 && i < 30000000 && j < 30000000) {
+        if (i >= -org.yatopiamc.yatopia.server.YatopiaConfig.biggestInt && j >= -org.yatopiamc.yatopia.server.YatopiaConfig.biggestInt && i < org.yatopiamc.yatopia.server.YatopiaConfig.biggestInt && j < org.yatopiamc.yatopia.server.YatopiaConfig.biggestInt) { // Yatopia - Remove world border limitations
             if (this.isChunkLoaded(i >> 4, j >> 4)) {
                 k = this.getChunkAt(i >> 4, j >> 4).getHighestBlock(heightmap_type, i & 15, j & 15) + 1;
             } else {
diff --git a/src/main/java/net/minecraft/server/WorldBorder.java b/src/main/java/net/minecraft/server/WorldBorder.java
index 6b4c007550c3e2e405314d8931630557655ce6f1..1477dc21e884dff75455f7914bd4b3cb9d03e07b 100644
--- a/src/main/java/net/minecraft/server/WorldBorder.java
+++ b/src/main/java/net/minecraft/server/WorldBorder.java
@@ -17,9 +17,9 @@ public class WorldBorder {
     private int f = 5;
     private double g;
     private double h;
-    private int i = 29999984;
-    private WorldBorder.a j = new WorldBorder.d(6.0E7D);
-    public static final WorldBorder.c c = new WorldBorder.c(0.0D, 0.0D, 0.2D, 5.0D, 5, 15, 6.0E7D, 0L, 0.0D);
+    private int i = org.yatopiamc.yatopia.server.YatopiaConfig.biggestInt; // Yatopia - Remove world border limitations
+    private WorldBorder.a j = new WorldBorder.d(org.yatopiamc.yatopia.server.YatopiaConfig.WborderSize); // Yatopia - Remove world border limitations
+    public static final WorldBorder.c c = new WorldBorder.c(0.0D, 0.0D, 0.2D, 5.0D, 5, 15, org.yatopiamc.yatopia.server.YatopiaConfig.WborderSize, 0L, 0.0D); // Yatopia - Remove world border limitations
     public WorldServer world; // CraftBukkit
 
     public WorldBorder() {}
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 300139b0a08a83647f6e010191c475cc65106a58..7143be96ec54fba60d2054ebed76640986223758 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -655,7 +655,7 @@ public final class CraftServer implements Server {
 
     @Override
     public int getMaxWorldSize() {
-        return this.getProperties().maxWorldSize;
+        return org.yatopiamc.yatopia.server.YatopiaConfig.biggestInt; // Yatopia - Remove world border limitations
     }
 
     @Override
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorldBorder.java b/src/main/java/org/bukkit/craftbukkit/CraftWorldBorder.java
index 044d46ac87f9e82b54ebf69b3cfd5842d5388d66..063f2f6d17c05d6ec2cf816c64ce045bec4b6fce 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorldBorder.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorldBorder.java
@@ -18,7 +18,7 @@ public class CraftWorldBorder implements WorldBorder {
 
     @Override
     public void reset() {
-        this.setSize(6.0E7D);
+        this.setSize(org.yatopiamc.yatopia.server.YatopiaConfig.WborderSize); // Yatopia - Remove world border limitations
         this.setDamageAmount(0.2D);
         this.setDamageBuffer(5.0D);
         this.setWarningDistance(5);
@@ -39,7 +39,7 @@ public class CraftWorldBorder implements WorldBorder {
     @Override
     public void setSize(double newSize, long time) {
         // PAIL: TODO: Magic Values
-        newSize = Math.min(6.0E7D, Math.max(1.0D, newSize));
+        newSize = Math.min(org.yatopiamc.yatopia.server.YatopiaConfig.WborderSize, Math.max(1.0D, newSize)); // Yatopia - Remove world border limitations
         time = Math.min(9223372036854775L, Math.max(0L, time));
 
         if (time > 0L) {
@@ -60,8 +60,8 @@ public class CraftWorldBorder implements WorldBorder {
     @Override
     public void setCenter(double x, double z) {
         // PAIL: TODO: Magic Values
-        x = Math.min(3.0E7D, Math.max(-3.0E7D, x));
-        z = Math.min(3.0E7D, Math.max(-3.0E7D, z));
+        x = Math.min(org.yatopiamc.yatopia.server.YatopiaConfig.replace2, Math.max(-org.yatopiamc.yatopia.server.YatopiaConfig.replace2, x)); // Yatopia - Remove world border limitations
+        z = Math.min(org.yatopiamc.yatopia.server.YatopiaConfig.replace2, Math.max(-org.yatopiamc.yatopia.server.YatopiaConfig.replace2, z)); // Yatopia - Remove world border limitations
 
         this.handle.setCenter(x, z);
     }
diff --git a/src/main/java/org/yatopiamc/yatopia/server/YatopiaConfig.java b/src/main/java/org/yatopiamc/yatopia/server/YatopiaConfig.java
index 15e2fa125bc293b954cceb5b1fbcec7fade3e4db..4e0b1cef88060a6d4063c92d49ba84959b0974ab 100644
--- a/src/main/java/org/yatopiamc/yatopia/server/YatopiaConfig.java
+++ b/src/main/java/org/yatopiamc/yatopia/server/YatopiaConfig.java
@@ -276,4 +276,19 @@ public class YatopiaConfig {
         fixProtocolLib = getBoolean("settings.fix-protocollib", fixProtocolLib);
     }
 
+    public static double WborderSize = 6.0E7D; // Vanilla 6.0E7D; Max: 4.294967294E9D // replacement for 6.0E7D
+    public static int biggestInt = 29999984; // Vanilla: 29999984; Max: 2147483647
+    private static void world() {
+        WborderSize = getDouble("settings.experimental.world.world-border-size", WborderSize);
+        biggestInt = getInt("settings.experimental.world.world-max-int", biggestInt);
+
+        // Start - World size settings checks
+        if (WborderSize > 4.294967294E9D) Bukkit.getLogger().log(Level.SEVERE, "Yatopia: settings.experimental.world.world-border-size is set over it's maximum value");
+        if (biggestInt > 2147483647) Bukkit.getLogger().log(Level.SEVERE, "Yatopia: settings.experimental.world.world-max-int is set over it's maximum value");
+        // End - World size settings checks
+    }
+    public static float biggestFloat = (float)WborderSize;
+    public static double replace1 = ((double)biggestInt)-112.0D; // replacement for 2.9999872E7D
+    public static double replace2 = WborderSize/2; // replacement for 3.0E7D
+    public static int replace3 = biggestInt + 16; // replacement for 30000000
 }
