From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Simon Gardling <titaniumtown@gmail.com>
Date: Fri, 12 Feb 2021 16:20:23 -0500
Subject: [PATCH] replace Math.sin and Math.cos with MathHelper


diff --git a/src/main/java/net/minecraft/server/EnderDragonBattle.java b/src/main/java/net/minecraft/server/EnderDragonBattle.java
index 0d8fba494ed11ad79201dfd1c7f3ad5b288ca0ca..48dd107f39572c33ff99a8e1362ae801930823d5 100644
--- a/src/main/java/net/minecraft/server/EnderDragonBattle.java
+++ b/src/main/java/net/minecraft/server/EnderDragonBattle.java
@@ -387,8 +387,8 @@ public class EnderDragonBattle {
     private void n() {
         if (!this.gateways.isEmpty()) {
             int i = (Integer) this.gateways.remove(this.gateways.size() - 1);
-            int j = MathHelper.floor(96.0D * Math.cos(2.0D * (-3.141592653589793D + 0.15707963267948966D * (double) i)));
-            int k = MathHelper.floor(96.0D * Math.sin(2.0D * (-3.141592653589793D + 0.15707963267948966D * (double) i)));
+            int j = MathHelper.floor(96.0D * MathHelper.cos(2.0D * (-3.141592653589793D + 0.15707963267948966D * (double) i))); // Yatopia
+            int k = MathHelper.floor(96.0D * MathHelper.sin(2.0D * (-3.141592653589793D + 0.15707963267948966D * (double) i))); // Yatopia
 
             this.a(new BlockPosition(j, 75, k));
         }
diff --git a/src/main/java/net/minecraft/server/EntityBoat.java b/src/main/java/net/minecraft/server/EntityBoat.java
index fcb4936e78d56907b0e56d0aaf7f8361513d0a4e..068095faeb42874a347ee1e6628dd304cca6762e 100644
--- a/src/main/java/net/minecraft/server/EntityBoat.java
+++ b/src/main/java/net/minecraft/server/EntityBoat.java
@@ -365,7 +365,7 @@ public class EntityBoat extends Entity {
 
             this.aC = MathHelper.a(this.aC, 0.0F, 1.0F);
             this.aE = this.aD;
-            this.aD = 10.0F * (float) Math.sin((double) (0.5F * (float) this.world.getTime())) * this.aC;
+            this.aD = 10.0F * (float) MathHelper.sin((double) (0.5F * (float) this.world.getTime())) * this.aC; // Yatopia
         } else {
             if (!this.aA) {
                 this.d(0);
diff --git a/src/main/java/net/minecraft/server/EntityEnderSignal.java b/src/main/java/net/minecraft/server/EntityEnderSignal.java
index 132fa358b844a609a9ede3032e98f991c5eac65a..b7e60a672d67b083253d68fc6addf93de4f47f6e 100644
--- a/src/main/java/net/minecraft/server/EntityEnderSignal.java
+++ b/src/main/java/net/minecraft/server/EntityEnderSignal.java
@@ -91,7 +91,7 @@ public class EntityEnderSignal extends Entity {
 
             int i = this.locY() < this.targetY ? 1 : -1;
 
-            vec3d = new Vec3D(Math.cos((double) f2) * d5, d6 + ((double) i - d6) * 0.014999999664723873D, Math.sin((double) f2) * d5);
+            vec3d = new Vec3D(MathHelper.cos((double) f2) * d5, d6 + ((double) i - d6) * 0.014999999664723873D, MathHelper.sin((double) f2) * d5); // Yatopia
             this.setMot(vec3d);
         }
 
diff --git a/src/main/java/net/minecraft/server/EntityFox.java b/src/main/java/net/minecraft/server/EntityFox.java
index ff78bd607f19a1b9f37281946db3f705c844f7cc..29035d14656393faddc82936dbf2dbe04f492f49 100644
--- a/src/main/java/net/minecraft/server/EntityFox.java
+++ b/src/main/java/net/minecraft/server/EntityFox.java
@@ -1094,8 +1094,8 @@ public class EntityFox extends EntityAnimal {
         private void j() {
             double d0 = 6.283185307179586D * EntityFox.this.getRandom().nextDouble();
 
-            this.c = Math.cos(d0);
-            this.d = Math.sin(d0);
+            this.c = MathHelper.cos(d0); // Yatopia
+            this.d = MathHelper.sin(d0); // Yatopia
             this.e = 80 + EntityFox.this.getRandom().nextInt(20);
         }
     }
diff --git a/src/main/java/net/minecraft/server/EntityGuardian.java b/src/main/java/net/minecraft/server/EntityGuardian.java
index 476f692df79e98779271cc5a395276ab9c755a58..75b89fbbc8bce25a8a0edebe6a6d0d2e26281f3a 100644
--- a/src/main/java/net/minecraft/server/EntityGuardian.java
+++ b/src/main/java/net/minecraft/server/EntityGuardian.java
@@ -362,10 +362,10 @@ public class EntityGuardian extends EntityMonster {
                 float f2 = MathHelper.g(0.125F, this.i.dN(), f1);
 
                 this.i.q(f2);
-                double d4 = Math.sin((double) (this.i.ticksLived + this.i.getId()) * 0.5D) * 0.05D;
-                double d5 = Math.cos((double) (this.i.yaw * 0.017453292F));
-                double d6 = Math.sin((double) (this.i.yaw * 0.017453292F));
-                double d7 = Math.sin((double) (this.i.ticksLived + this.i.getId()) * 0.75D) * 0.05D;
+                double d4 = MathHelper.sin((double) (this.i.ticksLived + this.i.getId()) * 0.5D) * 0.05D; // Yatopia
+                double d5 = MathHelper.cos((double) (this.i.yaw * 0.017453292F)); // Yatopia
+                double d6 = MathHelper.sin((double) (this.i.yaw * 0.017453292F)); // Yatopia
+                double d7 = MathHelper.sin((double) (this.i.ticksLived + this.i.getId()) * 0.75D) * 0.05D; // Yatopia
 
                 this.i.setMot(this.i.getMot().add(d4 * d5, d7 * (d6 + d5) * 0.25D + (double) f2 * d2 * 0.1D, d4 * d6));
                 ControllerLook controllerlook = this.i.getControllerLook();
diff --git a/src/main/java/net/minecraft/server/EntityHuman.java b/src/main/java/net/minecraft/server/EntityHuman.java
index 1730108aa67802ed15407c45542b6223ca86634f..348d7c093014ed4e26c6c08553892aaeb8c38dad 100644
--- a/src/main/java/net/minecraft/server/EntityHuman.java
+++ b/src/main/java/net/minecraft/server/EntityHuman.java
@@ -610,7 +610,7 @@ public abstract class EntityHuman extends EntityLiving {
                 float f5 = this.random.nextFloat() * 6.2831855F;
                 float f6 = 0.02F * this.random.nextFloat();
 
-                entityitem.setMot((double) (-f3 * f2 * 0.3F) + Math.cos((double) f5) * (double) f6, (double) (-f1 * 0.3F + 0.1F + (this.random.nextFloat() - this.random.nextFloat()) * 0.1F), (double) (f4 * f2 * 0.3F) + Math.sin((double) f5) * (double) f6);
+                entityitem.setMot((double) (-f3 * f2 * 0.3F) + MathHelper.cos((double) f5) * (double) f6, (double) (-f1 * 0.3F + 0.1F + (this.random.nextFloat() - this.random.nextFloat()) * 0.1F), (double) (f4 * f2 * 0.3F) + MathHelper.sin((double) f5) * (double) f6); // Yatopia
             }
 
             // CraftBukkit start - fire PlayerDropItemEvent
diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index ee27b814e0ae4eb9867df7fef31671fb23193cb7..81b3082a4499b983dbbd68a98fa80ba91a409662 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -160,15 +160,15 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
     // Paper start
     public BlockPosition getPointInFront(double inFront) {
         double rads = Math.toRadians(MCUtil.normalizeYaw(this.yaw+90)); // MC rotates yaw 90 for some odd reason
-        final double x = locX() + inFront * Math.cos(rads);
-        final double z = locZ() + inFront * Math.sin(rads);
+        final double x = locX() + inFront * MathHelper.cos(rads); // Yatopia
+        final double z = locZ() + inFront * MathHelper.sin(rads); // Yatopia
         return new BlockPosition(x, locY(), z);
     }
 
     public ChunkCoordIntPair getChunkInFront(double inFront) {
         double rads = Math.toRadians(MCUtil.normalizeYaw(this.yaw+90)); // MC rotates yaw 90 for some odd reason
-        final double x = locX() + (inFront * 16) * Math.cos(rads);
-        final double z = locZ() + (inFront * 16) * Math.sin(rads);
+        final double x = locX() + (inFront * 16) * MathHelper.cos(rads); // Yatopia
+        final double z = locZ() + (inFront * 16) * MathHelper.sin(rads); // Yatopia
         return new ChunkCoordIntPair(MathHelper.floor(x) >> 4, MathHelper.floor(z) >> 4);
     }
     // Paper end
diff --git a/src/main/java/net/minecraft/server/EntityRavager.java b/src/main/java/net/minecraft/server/EntityRavager.java
index 3606440e41ab3193fadebb455c0967dbc0ccbb63..b160e0886d431be7b281fd3d7d30e822cd26234c 100644
--- a/src/main/java/net/minecraft/server/EntityRavager.java
+++ b/src/main/java/net/minecraft/server/EntityRavager.java
@@ -182,9 +182,9 @@ public class EntityRavager extends EntityRaider {
 
     private void eX() {
         if (this.random.nextInt(6) == 0) {
-            double d0 = this.locX() - (double) this.getWidth() * Math.sin((double) (this.aA * 0.017453292F)) + (this.random.nextDouble() * 0.6D - 0.3D);
+            double d0 = this.locX() - (double) this.getWidth() * MathHelper.sin((double) (this.aA * 0.017453292F)) + (this.random.nextDouble() * 0.6D - 0.3D); // Yatopia
             double d1 = this.locY() + (double) this.getHeight() - 0.3D;
-            double d2 = this.locZ() + (double) this.getWidth() * Math.cos((double) (this.aA * 0.017453292F)) + (this.random.nextDouble() * 0.6D - 0.3D);
+            double d2 = this.locZ() + (double) this.getWidth() * MathHelper.cos((double) (this.aA * 0.017453292F)) + (this.random.nextDouble() * 0.6D - 0.3D); // Yatopia
 
             this.world.addParticle(Particles.ENTITY_EFFECT, d0, d1, d2, 0.4980392156862745D, 0.5137254901960784D, 0.5725490196078431D);
         }
diff --git a/src/main/java/net/minecraft/server/EntitySquid.java b/src/main/java/net/minecraft/server/EntitySquid.java
index 918cd774bc7e9f84ddbc1e65b4cb4d661eb127a5..d35ac0a170e4213be007147e3674645455feb3cd 100644
--- a/src/main/java/net/minecraft/server/EntitySquid.java
+++ b/src/main/java/net/minecraft/server/EntitySquid.java
@@ -44,8 +44,8 @@ public class EntitySquid extends EntityWaterAnimal {
 
     private void rotateVectorAroundY(org.bukkit.util.Vector vector, double degrees) {
         double rad = Math.toRadians(degrees);
-        double cos = Math.cos(rad);
-        double sine = Math.sin(rad);
+        double cos = MathHelper.cos(rad); // Yatopia
+        double sine = MathHelper.sin(rad); // Yatopia
         double x = vector.getX();
         double z = vector.getZ();
         vector.setX(cos * x - sine * z);
diff --git a/src/main/java/net/minecraft/server/EntityTNTPrimed.java b/src/main/java/net/minecraft/server/EntityTNTPrimed.java
index 6e77dc89f5441f0f483571fee9aa9f34b6d1dd1c..9271efb508a1b73fff8dd7f5d8c685a9f92d6a35 100644
--- a/src/main/java/net/minecraft/server/EntityTNTPrimed.java
+++ b/src/main/java/net/minecraft/server/EntityTNTPrimed.java
@@ -24,7 +24,7 @@ public class EntityTNTPrimed extends Entity {
         this.setPosition(d0, d1, d2);
         double d3 = world.random.nextDouble() * 6.2831854820251465D;
 
-        this.setMot(-Math.sin(d3) * 0.02D, 0.20000000298023224D, -Math.cos(d3) * 0.02D);
+        this.setMot(-MathHelper.sin(d3) * 0.02D, 0.20000000298023224D, -MathHelper.cos(d3) * 0.02D); // Yatopia
         this.setFuseTicks(80);
         this.lastX = d0;
         this.lastY = d1;
diff --git a/src/main/java/net/minecraft/server/MathHelper.java b/src/main/java/net/minecraft/server/MathHelper.java
index 1b7cc2fc9a9636a245c5570e2e326973129d095b..770fe8bf9bad092357bf24361acab98516d188aa 100644
--- a/src/main/java/net/minecraft/server/MathHelper.java
+++ b/src/main/java/net/minecraft/server/MathHelper.java
@@ -29,6 +29,14 @@ public class MathHelper {
         return CompactSineLUT.cos(f); // Yatopia - CompactSineLUT
     }
 
+    public static float sin(double f) {
+        return CompactSineLUT.sin((float)f); // Yatopia - CompactSineLUT
+    }
+
+    public static float cos(double f) {
+        return CompactSineLUT.cos((float)f); // Yatopia - CompactSineLUT
+    }
+
     public static float c(float f) {
         return (float) Math.sqrt((double) f);
     }
diff --git a/src/main/java/net/minecraft/server/RandomPositionGenerator.java b/src/main/java/net/minecraft/server/RandomPositionGenerator.java
index aaa89a39bf697bd0f3354005b428d858f065f66b..1f6606165a45c30eddca8d06ea14067b0bb09f8b 100644
--- a/src/main/java/net/minecraft/server/RandomPositionGenerator.java
+++ b/src/main/java/net/minecraft/server/RandomPositionGenerator.java
@@ -155,8 +155,8 @@ public class RandomPositionGenerator {
             double d1 = MathHelper.d(vec3d.z, vec3d.x) - 1.5707963705062866D;
             double d2 = d1 + (double) (2.0F * random.nextFloat() - 1.0F) * d0;
             double d3 = Math.sqrt(random.nextDouble()) * (double) MathHelper.a * (double) i;
-            double d4 = -d3 * Math.sin(d2);
-            double d5 = d3 * Math.cos(d2);
+            double d4 = -d3 * MathHelper.sin(d2); // Yatopia
+            double d5 = d3 * MathHelper.cos(d2); // Yatopia
 
             if (Math.abs(d4) <= (double) i && Math.abs(d5) <= (double) i) {
                 int l = random.nextInt(2 * j + 1) - j + k;
