From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Sotr <i@omc.hk>
Date: Wed, 15 Apr 2020 22:17:18 +0700
Subject: [PATCH] (Akarin) Disable the Snooper


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 45f6c0e295aca017763fa9502644773cd04d428e..b20bb542c19ec3e19b5c98a0ef1ab78f7724399a 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1462,6 +1462,8 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
         //} // Paper
 
         this.methodProfiler.enter("snooper");
+        // Akarin - Disable Snooper
+        /*
         if (((DedicatedServer) this).getDedicatedServerProperties().snooperEnabled && !this.snooper.d() && this.ticks > 100) { // Spigot
             this.snooper.a();
         }
@@ -1469,6 +1471,7 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
         if (((DedicatedServer) this).getDedicatedServerProperties().snooperEnabled && this.ticks % 6000 == 0) { // Spigot
             this.snooper.b();
         }
+        */
 
         this.methodProfiler.exit();
 
