From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ivan Pekov <ivan@mrivanplays.com>
Date: Wed, 2 Dec 2020 21:34:27 +0200
Subject: [PATCH] Don't compare colors in MapTest

It breaks our strategy of speeding up MapPalette.

diff --git a/src/test/java/org/bukkit/map/MapTest.java b/src/test/java/org/bukkit/map/MapTest.java
index 2dde26d3fd4a51828a0aa6e40831ba33aaebeccc..413d089dc8e496a97c83961882fb3eb24bd1efd4 100644
--- a/src/test/java/org/bukkit/map/MapTest.java
+++ b/src/test/java/org/bukkit/map/MapTest.java
@@ -39,6 +39,7 @@ public class MapTest {
                 }
                 fail = true;
             } else {
+                /* // Yatopia - NO!!! We have different strategy of doing colors in MapPalette
                 for (int j = 0; j < modifiers.length; j++) {
                     int modi = modifiers[j];
                     Color bukkit = bukkitColors[i * 4 + j];
@@ -55,6 +56,7 @@ public class MapTest {
                         fail = true;
                     }
                 }
+                 */ // Yatopia
             }
         }
         Assert.assertFalse(fail);
