From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: The Arcane Brony <myrainbowdash949@gmail.com>
Date: Mon, 24 May 2021 06:07:10 +0200
Subject: [PATCH] Momiji best awoo


diff --git a/src/main/java/net/minecraft/server/EULA.java b/src/main/java/net/minecraft/server/EULA.java
index b6d94005cb2d7f28fc4f0b78191ef1ba13190e56..af13f50817f5c701c4e7331711580e5c999e5219 100644
--- a/src/main/java/net/minecraft/server/EULA.java
+++ b/src/main/java/net/minecraft/server/EULA.java
@@ -83,9 +83,8 @@ public class EULA {
 
                 try {
                     Properties properties = new Properties();
-
                     properties.setProperty("eula", "false");
-                    properties.store(outputstream, "By changing the setting below to TRUE you are indicating your agreement to our EULA (https://account.mojang.com/documents/minecraft_eula).\nYou also agree that tacos are tasty, and the best food in the world.");  // Paper - fix lag; // Tuinity - Tacos are disgusting // Purpur - no they're not
+                    properties.store(outputstream, "By changing the setting below to TRUE you are indicating your agreement to our EULA (https://account.mojang.com/documents/minecraft_eula).\nYou also agree that tacos are tasty, and the best food in the world.\nIn addition, you also agree that Momiji Inubashiri is best awoo!");  // Paper - fix lag; // Tuinity - Tacos are disgusting // Purpur - no they're not // TheArcaneBrony @ Yatopia - Momiji best awoo! AWOOOOO!~
                 } catch (Throwable throwable1) {
                     throwable = throwable1;
                     throw throwable1;
