From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: The Arcane Brony <myrainbowdash949@gmail.com>
Date: Mon, 24 May 2021 06:11:31 +0200
Subject: [PATCH] Clean up some code


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 61d7004e51fcfb5568750665065261295cacd2f7..ce537041bbc989a04d39cf4a665fb4f4a08363e4 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -393,7 +393,6 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
                         this.a = SystemUtils.getMonotonicMillis();
                         MinecraftServer.LOGGER.info("Converting... {}%", i);
                     }
-
                 }
 
                 @Override
diff --git a/src/main/java/net/minecraft/server/gui/GuiStatsComponent.java b/src/main/java/net/minecraft/server/gui/GuiStatsComponent.java
index 79cc587d8ad9f576f9e4478a41baabcaf690c568..6ceec044398c73a0a405863e16816d2514d9bfc5 100644
--- a/src/main/java/net/minecraft/server/gui/GuiStatsComponent.java
+++ b/src/main/java/net/minecraft/server/gui/GuiStatsComponent.java
@@ -48,7 +48,6 @@ public class GuiStatsComponent extends JComponent {
         this.d[1] = "Avg tick: " + GuiStatsComponent.a.format(this.a(this.e.h) * 1.0E-6D) + " ms";
         getStatEntries()[2] = "TPS from last 1m, 5m, 15m: " + String.join(", ", tpsAvg);
         // Paper end
-
         this.b[this.c++ & 255] = (int) (i * 100L / Runtime.getRuntime().maxMemory());
         this.repaint();
     }
diff --git a/src/main/java/net/minecraft/server/gui/ServerGUI.java b/src/main/java/net/minecraft/server/gui/ServerGUI.java
index d1d98c2f192514a3f511bebb00c088b03e6be84c..0acecb53031c6fbd1e5bf3a10a0a22dfee4b1730 100644
--- a/src/main/java/net/minecraft/server/gui/ServerGUI.java
+++ b/src/main/java/net/minecraft/server/gui/ServerGUI.java
@@ -12,16 +12,7 @@ import java.awt.event.WindowAdapter;
 import java.awt.event.WindowEvent;
 import java.util.Collection;
 import java.util.concurrent.atomic.AtomicBoolean;
-import javax.swing.JComponent;
-import javax.swing.JFrame;
-import javax.swing.JList;
-import javax.swing.JPanel;
-import javax.swing.JScrollBar;
-import javax.swing.JScrollPane;
-import javax.swing.JTextArea;
-import javax.swing.JTextField;
-import javax.swing.SwingUtilities;
-import javax.swing.UIManager;
+import javax.swing.*;
 import javax.swing.border.EtchedBorder;
 import javax.swing.border.TitledBorder;
 import javax.swing.text.AttributeSet;
@@ -73,7 +64,7 @@ public class ServerGUI extends JComponent {
         jframe.addWindowListener(new WindowAdapter() {
             public void windowClosing(WindowEvent windowevent) {
                 if (!servergui.f.getAndSet(true)) {
-                    jframe.setTitle("Minecraft server - shutting down!");
+                    jframe.setTitle(jframe.getTitle() + " - shutting down!");
                     dedicatedserver.safeShutdown(true);
                     servergui.f();
                 }
