From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Simon Gardling <titaniumtown@gmail.com>
Date: Mon, 1 Mar 2021 11:34:52 -0500
Subject: [PATCH] fix Purpur rideables patch


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 8735735eb116382b21b6751108629acb9411d332..ab5fb00885d22126639e7459ef71c56be4b5eb9e 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -177,6 +177,7 @@ public abstract class Entity implements INamableTileEntity, ICommandListener, ne
     private CraftEntity bukkitEntity;
 
     public PlayerChunkMap.EntityTracker tracker; // Paper package private -> public
+    public PlayerChunkMap.EntityTracker getTracker() { return tracker; } // Yatopia
     public boolean collisionLoadChunks = false; // Paper
     public Throwable addedToWorldStack; // Paper - entity debug
     public CraftEntity getBukkitEntity() {
diff --git a/src/main/java/net/pl3x/purpur/controller/ControllerLookWASD.java b/src/main/java/net/pl3x/purpur/controller/ControllerLookWASD.java
index 292821f039d99a03ba4daeb3a941616ef5f6287e..bc56aeadaf32f1e6f17da3ab285f6c161880fbba 100644
--- a/src/main/java/net/pl3x/purpur/controller/ControllerLookWASD.java
+++ b/src/main/java/net/pl3x/purpur/controller/ControllerLookWASD.java
@@ -42,7 +42,7 @@ public class ControllerLookWASD extends ControllerLook {
         entity.setHeadRotation(entity.yaw);
         entity.pitch = normalizePitch(pitch + pitchOffset);
 
-        entity.tracker.broadcast(new PacketPlayOutEntity
+        entity.getTracker().broadcast(new PacketPlayOutEntity
                 .PacketPlayOutRelEntityMoveLook(entity.getId(),
                 (short) 0, (short) 0, (short) 0,
                 (byte) MathHelper.d(entity.yaw * 256.0F / 360.0F),
